<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<!-- 2025-01-29 Wed 13:42 -->
<meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<title>BEAST2 notes</title>
<meta name="author" content="Alexander E. Zarebski" />
<meta name="generator" content="Org Mode" />
<style>
  #content { max-width: 60em; margin: auto; }
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #e6e6e6;
    border-radius: 3px;
    background-color: #f2f2f2;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: auto;
  }
  pre.src:before {
    display: none;
    position: absolute;
    top: -8px;
    right: 12px;
    padding: 3px;
    color: #555;
    background-color: #f2f2f299;
  }
  pre.src:hover:before { display: inline; margin-top: 14px;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-authinfo::before { content: 'Authinfo'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .equation-container {
    display: table;
    text-align: center;
    width: 100%;
  }
  .equation {
    vertical-align: middle;
  }
  .equation-label {
    display: table-cell;
    text-align: right;
    vertical-align: middle;
  }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { }
</style>
<link rel="icon" type="image/png" href="../resources/nicemacs-favicon.png">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.1/normalize.css">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/milligram/1.4.1/milligram.css">
<link rel="stylesheet" href="../microgram.css">
<script>
  window.MathJax = {
    tex: {
      ams: {
        multlineWidth: '85%'
      },
      tags: 'ams',
      tagSide: 'right',
      tagIndent: '.8em'
    },
    chtml: {
      scale: 1.0,
      displayAlign: 'center',
      displayIndent: '0em'
    },
    svg: {
      scale: 1.0,
      displayAlign: 'center',
      displayIndent: '0em'
    },
    output: {
      font: 'mathjax-modern',
      displayOverflow: 'overflow'
    }
  };
</script>

<script
  id="MathJax-script"
  async
  src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js">
</script>
</head>
<body>
<div id="content" class="content">
<h1 class="title">BEAST2 notes</h1>
<p>
<a href="../index.html">Home</a>
</p>


<div id="orgbb46b11" class="figure">
<p><img src="../images/beast2-logo.png" alt="beast2-logo.png" width="250px" />
</p>
</div>

<div id="table-of-contents" role="doc-toc">
<h2>Table of Contents</h2>
<div id="text-table-of-contents" role="doc-toc">
<ul>
<li><a href="#org303831a">Links</a></li>
<li><a href="#org072bb36">Notes</a>
<ul>
<li><a href="#org589e7e4">Understanding BEAUti</a></li>
<li><a href="#org7af489c">Build a BEAST2</a></li>
<li><a href="#orge6e1748"><span class="todo TODO">TODO</span> Understanding Packages</a></li>
<li><a href="#orga300d01">Understanding BEAST2</a></li>
<li><a href="#org46645d3">Understanding the XML</a></li>
<li><a href="#orgbb56c34">Understanding the MCMC application</a></li>
</ul>
</li>
<li><a href="#orge446225">Simulating data</a>
<ul>
<li><a href="#org559ab4e">ReMASTER</a></li>
</ul>
</li>
<li><a href="#org7fbda54">Tools and useful functions</a>
<ul>
<li><a href="#orgd344950">Cogsworth</a></li>
<li><a href="#org4f66aff">Cogsworth 2.0: snakemake flavoured Cogsworth</a></li>
<li><a href="#org932135b">Web beautify</a></li>
<li><a href="#sec:read-beast2-log">Read BEAST2 log file</a></li>
<li><a href="#org58dfa98">Running BEAST2 XML files with Python</a></li>
<li><a href="#orgb0431cf">Pretty printing Tracer output</a></li>
</ul>
</li>
<li><a href="#org2697552">Examples</a>
<ul>
<li><a href="#orgbd9ccea">BEAST2 as an MCMC engine: part I</a></li>
<li><a href="#org2abb9f1">BEAST2 as an MCMC engine: part II</a></li>
<li><a href="#org7ecb81e">Birth-death simulation: Example I</a></li>
<li><a href="#orga2df028">Birth-death simulation: Example II</a></li>
<li><a href="#orgb779836">Simulating sequences</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-org303831a" class="outline-2">
<h2 id="org303831a">Links</h2>
<div class="outline-text-2" id="text-org303831a">
<ul class="org-ul">
<li><a href="https://linguaphylo.github.io/about/">LinguaPhylo</a> is a scripting language that can be transpiled to BEAST2 XML</li>
<li><a href="https://www.beast2.org/xml/index.html">BEAST2 Documentation</a> for <i>the user who wants to edit BEAST2 XML manually.</i></li>
<li><a href="https://github.com/CompEvol/beast2">BEAST2 repository</a></li>
<li><a href="https://github.com/tgvaughan/feast">feast</a> a collection of resources from Tim Vaughan</li>
<li><a href="https://taming-the-beast.org/">Taming the BEAST</a> is a collection of tutorials for BEAST2</li>
</ul>
</div>
</div>

<div id="outline-container-org072bb36" class="outline-2">
<h2 id="org072bb36">Notes</h2>
<div class="outline-text-2" id="text-org072bb36">
</div>
<div id="outline-container-org589e7e4" class="outline-3">
<h3 id="org589e7e4">Understanding BEAUti</h3>
<div class="outline-text-3" id="text-org589e7e4">
<ul class="org-ul">
<li>The configuration of BEAUti is done using <i>template</i> files. There are a couple
of blog posts intended to make it easier to write these files:
<ul class="org-ul">
<li><a href="https://www.beast2.org/2015/06/16/better-beauti-templates.html">Better BEAUti Templates</a></li>
<li><a href="https://www.beast2.org/2022/02/01/debugging-beauti-templates.html">Debugging BEAUti Templates</a></li>
</ul></li>
</ul>
</div>
</div>

<div id="outline-container-org7af489c" class="outline-3">
<h3 id="org7af489c">Build a BEAST2</h3>
<div class="outline-text-3" id="text-org7af489c">
<ol class="org-ol">
<li>Clone the BEAST2 repository.
<ul class="org-ul">
<li><code>git clone git@github.com:CompEvol/beast2.git</code></li>
</ul></li>
<li>Build the project by running <code>ant</code>.
<ul class="org-ul">
<li>If you don't care about running the tests <code>ant compile-all</code> does the
compilation without running the tests.</li>
</ul></li>
<li>Run the desired program:
<ul class="org-ul">
<li><code>java -cp build/dist/beast.jar beast.app.beauti.Beauti</code></li>
<li><code>java -cp build/dist/beast.jar beast.app.beastapp.BeastMain</code></li>
</ul></li>
</ol>

<p>
Note that you can clock on the little '?' in Beauti to get the location that
packages are stored at.
</p>
</div>

<div id="outline-container-orgce28351" class="outline-4">
<h4 id="orgce28351">Notes</h4>
<div class="outline-text-4" id="text-orgce28351">
<ul class="org-ul">
<li><code>v2.6.7</code>:</li>
</ul>
</div>
</div>
</div>

<div id="outline-container-orge6e1748" class="outline-3">
<h3 id="orge6e1748"><span class="todo TODO">TODO</span> Understanding Packages</h3>
<div class="outline-text-3" id="text-orge6e1748">
<p>
Bouckaert has a <a href="https://www.beast2.org/2021/06/21/building-from-source.html">blog post</a> about building package from source from mid-2021. The
following line from the Babel <code>build.xml</code> (which is the example package used in
the blog post)
</p>

<div class="org-src-container">
<pre class="src src-xml">&lt;property name="beast2path" location="../beast2" /&gt;
</pre>
</div>

<p>
makes it sound like it assumes that BEAST2 will be in the same directory as the
packages during compilation.
</p>
</div>

<div id="outline-container-org114210a" class="outline-4">
<h4 id="org114210a">Installing packages from source</h4>
<div class="outline-text-4" id="text-org114210a">
<p>
Remco has written about <a href="https://www.beast2.org/2021/06/21/building-from-source.html">building and installing packages from source</a>.
</p>

<p>
The key steps include the following:
</p>

<ol class="org-ol">
<li>Get the <code>package.zip</code> files, e.g., <code>BEASTlabs.v2.0.2.zip</code>
<ul class="org-ul">
<li>If you are working on a server, you can <code>wget</code> the <code>.zip</code> file
from GitHub.</li>
</ul></li>
<li>Find the <code>~/.beast/2.X/</code> directory and make sure there is an empty
subdirectory with the same name as the package, e.g.,
<code>~/.beast/2.X/BEASTlabs</code> should be empty.</li>
<li>Use <code>zip -o &lt;path/to/package.zip&gt;</code> from <b>within</b> the directory to
install the package.
<ul class="org-ul">
<li>If this doesn't work, try <code>unzip</code>.</li>
</ul></li>
</ol>
</div>
</div>

<div id="outline-container-org4076341" class="outline-4">
<h4 id="org4076341"><span class="todo TODO">TODO</span> Configuring IntelliJ for writing a package</h4>
<div class="outline-text-4" id="text-org4076341">
<p>
There are notes on Java and IntelliJ <a href="./java-notes.html">here</a>. I have no idea if this is
the correct way that BEAST2 packages should be set up, but this has
worked for me in the past, and provided there are a sufficient number
of unit tests I assume it should work&#x2026;
</p>

<ul class="org-ul">
<li>Note that the <code>build.xml</code> probably assumes that BEAST2 and your code are in
sibling directories.</li>
<li>Build the BEAST2 project first and then close it and open your project
(probably based on the results of <code>ant skeleton</code>).</li>
<li>Click through <b>File</b> &#x2013; <b>Project Settings</b> &#x2013; <b>Modules</b> &#x2013; <b>&lt;YOUR/PACKAGE&gt;</b> and then
using the <b>+</b> (add) button, tell IntelliJ that you want to add a Module
Dependency, which will be BEAST2.</li>
<li>When building the package, if it complains about missing packages, you may
need to add JAR files in the same way.</li>
</ul>
</div>
</div>
</div>

<div id="outline-container-orga300d01" class="outline-3">
<h3 id="orga300d01">Understanding BEAST2</h3>
<div class="outline-text-3" id="text-orga300d01">
<ul class="org-ul">
<li>A class diagram showing the relationship between the various classes that are
relevant to tree priors are shown in Figure <a href="#org8800801">2</a>.</li>
<li>The <code>BEASTInterface</code> is an interface which all BEAST-objects must implement, the
<code>BEASTObject</code> class implements that interface and most objects we care about
will derive from <code>BEASTObject</code>.</li>
<li>The <code>StateNode</code> is akin to the <i>stochastic node</i> of H\"{o}hna <i>et al</i> (2014), ie it
is a random variable. State nodes can be changed by operators.</li>
<li>The <code>CalculationNode</code> is akin to the <i>deterministic</i> node of H\"{o}hna <i>et al</i>
(2014), ie it is a deterministic function of its input.</li>
<li>See Figure <a href="#orga839dea">1</a> which is helpful for understanding where
state and calculations nodes sit in this formulation.</li>
<li>The <code>TreeDistribution</code> (which extends the <code>Distribution</code> abstract class) is the
foundation for both the classes deriving from <code>SpeciesTreeDistribution</code> (which
are the birth-death models) and various coalescent models.</li>
<li>There is a <code>ModelBuilder</code> application for visualising a computation.</li>
<li><code>Operator</code> s are used to change the value of <code>StateNode</code> (of which <code>RealParameter</code>
is an example).</li>
</ul>


<div id="orga839dea" class="figure">
<p><img src="../images/drummond2015bayesian-state-and-calc-nodes.png" alt="drummond2015bayesian-state-and-calc-nodes.png" width="900px" />
</p>
<p><span class="figure-number">Figure 1: </span>Figure from Drummond and Bouckaert (2015).</p>
</div>


<div id="org8800801" class="figure">
<p><img src="../resources/beast2p6p6-class-diagram.png" alt="beast2p6p6-class-diagram.png" width="900px" />
</p>
<p><span class="figure-number">Figure 2: </span>Class diagram relating to tree priors in BEASTv2.6</p>
</div>

<p>
In Fig <a href="#orgdeacd4a">3</a> we see API of a Tree.
</p>


<div id="orgdeacd4a" class="figure">
<p><img src="../misc/plantuml/beastv2p7-trees.png" alt="beastv2p7-trees.png" width="400px" />
</p>
<p><span class="figure-number">Figure 3: </span>Class diagram relating to the tree in BEASTv2.7</p>
</div>


<div id="org4bfdc17" class="figure">
<p><img src="../misc/plantuml/beastv2p7-mcmc.png" alt="beastv2p7-mcmc.png" width="400px" />
</p>
<p><span class="figure-number">Figure 4: </span>Class diagram relating to the MCMC in BEASTv2.7</p>
</div>
</div>

<div id="outline-container-org9da0dbf" class="outline-4">
<h4 id="org9da0dbf">Packages provided by BEAST2</h4>
<div class="outline-text-4" id="text-org9da0dbf">
</div>
<ul class="org-ul">
<li><a id="org63f32e9"></a>Inference<br />
<div class="outline-text-5" id="text-org63f32e9">
<p>
The <code>beast.inference.distribution</code> package (which previously was part of either
the <code>beast.core</code> or <code>beast.math</code> packages) contains the classes from the <a href="https://commons.apache.org/proper/commons-math/">Apache
Commons Mathematics Library</a>, (it appears to use version 2.x). Since the
triangular distribution was not added until 3.0 it is not part of BEAST2.
</p>
</div>
</li>
</ul>
</div>

<div id="outline-container-orgfb6876d" class="outline-4">
<h4 id="orgfb6876d">Understanding <code>bdsky</code></h4>
<div class="outline-text-4" id="text-orgfb6876d">
<p>
To understand the implementation of time varying parameters see the following:
</p>

<ul class="org-ul">
<li><a href="#org7e61c0a">Skyline (interface)</a></li>
<li><a href="#orgc6003eb">SkylineSegement (class)</a></li>
<li><a href="#org88e505d">SimpleSkyline (class)</a></li>
<li><a href="#org490fd28">MultiSkyline (class)</a></li>
</ul>

<p>
To understand how this implements the BDSky model see the following:
</p>

<ul class="org-ul">
<li><a href="#orgd1cc754">BDSSkylineSegment (class)</a></li>
<li><a href="#orgb07ed78">BDSParameterization (abstract class)</a></li>
<li><a href="#org4d5760d">CanonicalParameterization (class)</a></li>
<li><a href="#orgee6fa7c">R0Parameterization (class)</a></li>
</ul>

<p>
which all culminates in <a href="#org814d6f5">ParameterizedBirthDeathSkylineModel (class)</a> which is the
class that actually does the main calculation.
</p>
</div>

<ul class="org-ul">
<li><a id="org7e61c0a"></a>Skyline (interface)<br />
<div class="outline-text-5" id="text-org7e61c0a">
<div class="org-src-container">
<pre class="src src-java">public interface Skyline { }
</pre>
</div>

<p>
<a href="https://github.com/BEAST2-Dev/bdsky/blob/master/src/bdsky/Skyline.java"><code>Skyline</code></a> is an <i>interface</i> provided by the <code>bdsky</code> package
</p>

<ul class="org-ul">
<li>This represents a multivariate right continuous function on a domain
\([a,\infty)\).</li>
<li>The interface requires the methods <code>getSegments</code>, <code>getValue</code>, <code>getValues</code>, <code>getTimes</code>
and <code>getDimension</code>.</li>
<li>Things that implement <code>Skyline</code> look like lists of</li>
</ul>
</div>
</li>

<li><a id="orgc6003eb"></a>SkylineSegment (class)<br />
<div class="outline-text-5" id="text-orgc6003eb">
<div class="org-src-container">
<pre class="src src-java">public class SkylineSegment { }
</pre>
</div>

<p>
<a href="https://github.com/BEAST2-Dev/bdsky/blob/master/src/bdsky/SkylineSegment.java"><code>SkylineSegment</code></a> is a <i>class</i> provided by the <code>bdsky</code> package
</p>

<ul class="org-ul">
<li>This represents an interval of time, \([a,b)\), and a vector of values
assigned to that time.</li>
<li>There are methods <code>start</code> and <code>end</code> to get the start, \(a\), and end, \(b\) times
of the interval.</li>
</ul>
</div>
</li>

<li><a id="org88e505d"></a>SimpleSkyline (class)<br />
<div class="outline-text-5" id="text-org88e505d">
<div class="org-src-container">
<pre class="src src-java">public class SimpleSkyline extends CalculationNode implements Skyline { }
</pre>
</div>

<p>
<a href="https://github.com/BEAST2-Dev/bdsky/blob/master/src/bdsky/SimpleSkyline.java"><code>SimpleSkyline</code></a> is a <i>class</i> provided by the <code>bdsky</code> package which represents a
piece-wise constant function (with a one dimensional range).
</p>
</div>
</li>

<li><a id="org490fd28"></a>MultiSkyline (class)<br />
<div class="outline-text-5" id="text-org490fd28">
<div class="org-src-container">
<pre class="src src-java">public class MultiSkyline extends CalculationNode implements Skyline { }
</pre>
</div>

<p>
<a href="https://github.com/BEAST2-Dev/bdsky/blob/ec2caf66a58b59cc82a7b8391dbd06b25fe67ca6/src/bdsky/MultiSkyline.java"><code>MultiSkyline</code></a> is a <i>class</i> provided by the <code>bdsky</code> package which is really just a
wrapper around a list of <a href="#org88e505d">=SimpleSkyline=s</a>, which takes care of splitting up time
into the required number of segments.
</p>
</div>
</li>

<li><a id="orgd1cc754"></a>BDSSkylineSegment (class)<br />
<div class="outline-text-5" id="text-orgd1cc754">
<div class="org-src-container">
<pre class="src src-java">public class BDSSkylineSegment extends SkylineSegment { }
</pre>
</div>

<p>
<a href="https://github.com/BEAST2-Dev/bdsky/blob/ec2caf66a58b59cc82a7b8391dbd06b25fe67ca6/src/bdsky/BDSSkylineSegment.java#L6"><code>BDSSkylineSegment</code></a> is a class that wraps the <code>SkylineSegment</code> and specialises it
for use as the parameterisation of the BDS model. Both the
<code>CanonicalParameterization</code> and the <code>R0Parameterization</code> make use of this class,
constructing their skyline segments with it.
</p>
</div>
</li>

<li><a id="orgb07ed78"></a>BDSParameterization (abstract class)<br />
<div class="outline-text-5" id="text-orgb07ed78">
<div class="org-src-container">
<pre class="src src-java">public abstract class BDSParameterization extends CalculationNode { }
</pre>
</div>

<p>
<a href="https://github.com/BEAST2-Dev/bdsky/blob/ec2caf66a58b59cc82a7b8391dbd06b25fe67ca6/src/bdsky/BDSParameterization.java"><code>BDSParameterization</code></a> is an <i>abstract class</i> provided by the <code>bdsky</code> package which
provides a template for writing classes which provide novel parameterisations of
the model. The <code>CanonicalParamterization</code> and <code>R0Parameterization</code> are two classes
that inherit from this.
</p>
</div>
</li>

<li><a id="org4d5760d"></a>CanonicalParameterization (class)<br />
<div class="outline-text-5" id="text-org4d5760d">
<div class="org-src-container">
<pre class="src src-java">public class CanonicalParameterization extends BDSParameterization { }
</pre>
</div>

<p>
<a href="https://github.com/BEAST2-Dev/bdsky/blob/ec2caf66a58b59cc82a7b8391dbd06b25fe67ca6/src/bdsky/CanonicalParameterization.java"><code>CanonicalParameterization</code></a> is a <i>class</i> provided by the <code>bdsky</code> package which
extends (ie inherits from) the <code>BDSParameterization</code> abstract class and represents
the parameterisation in terms of the birth rate, death rate, sampling rate, and
removal probability (which is important for sampled ancestors).
</p>
</div>
</li>

<li><a id="orgee6fa7c"></a>R0Parameterization (class)<br />
<div class="outline-text-5" id="text-orgee6fa7c">
<div class="org-src-container">
<pre class="src src-java">public class R0Parameterization extends BDSParameterization { }
</pre>
</div>

<p>
<a href="https://github.com/BEAST2-Dev/bdsky/blob/ec2caf66a58b59cc82a7b8391dbd06b25fe67ca6/src/bdsky/R0Parameterization.java"><code>R0Parameterization</code></a> is a <i>class</i> provided by the <code>bdsky</code> package which extends (ie
inherits from) the <code>BDSParameterization</code> abstract class and represents the
parameterisation in terms of the reproductive number, total becoming
uninfectious rate, the sampling proportion and the removal probability upon
sampling.
</p>
</div>
</li>

<li><a id="org814d6f5"></a>ParameterizedBirthDeathSkylineModel (class)<br />
<div class="outline-text-5" id="text-org814d6f5">
<div class="org-src-container">
<pre class="src src-java">public class ParameterizedBirthDeathSkylineModel extends SpeciesTreeDistribution { }
</pre>
</div>

<p>
<a href="https://github.com/BEAST2-Dev/bdsky/blob/ec2caf66a58b59cc82a7b8391dbd06b25fe67ca6/src/bdsky/ParameterizedBirthDeathSkylineModel.java"><code>ParameterizedBirthDeathSkylineModel</code></a> is a <i>class</i> provided by the <code>bdsky</code> package
which does the actual calculation, ie it has a <code>calculateTreeLogLikelihood</code>
method.
</p>
</div>
</li>
</ul>
</div>
</div>

<div id="outline-container-org46645d3" class="outline-3">
<h3 id="org46645d3">Understanding the XML</h3>
<div class="outline-text-3" id="text-org46645d3">
<p>
Chapter 13 of (, ) focuses on how BEAST uses XML to
specify a computation.
</p>

<ul class="org-ul">
<li>There are some reserved attributes (<a href="#org1bda6f8">this table</a>) and element names (<a href="#org42382cf">this table</a>)
that BEAST attributes special meaning to.</li>
<li>The <code>namespace</code> attribute of the <code>beast</code> element specifies where classes are
looked for. The comma-separated list is searched in order, but you can use the
full name in cases where there could be a name-clash.</li>
<li>The <code>input</code> can be replaced with the <code>name</code> to simplify the specification.</li>
</ul>

<table id="org1bda6f8" border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">
<caption class="t-above"><span class="table-number">Table 1:</span> Reserved attributes</caption>

<colgroup>
<col  class="org-left" />

<col  class="org-left" />
</colgroup>
<thead>
<tr>
<th scope="col" class="org-left">Attribute name</th>
<th scope="col" class="org-left">Purpose</th>
</tr>
</thead>
<tbody>
<tr>
<td class="org-left"><code>id</code></td>
<td class="org-left">Unique identifier</td>
</tr>

<tr>
<td class="org-left"><code>idref</code></td>
<td class="org-left">Reference to another element</td>
</tr>

<tr>
<td class="org-left"><code>name</code></td>
<td class="org-left">The part of the parent it plugs into</td>
</tr>

<tr>
<td class="org-left"><code>spec</code></td>
<td class="org-left">The Java class to use</td>
</tr>
</tbody>
</table>

<table id="org42382cf" border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">
<caption class="t-above"><span class="table-number">Table 2:</span> Reserved tags</caption>

<colgroup>
<col  class="org-left" />

<col  class="org-left" />
</colgroup>
<thead>
<tr>
<th scope="col" class="org-left">Tag name</th>
<th scope="col" class="org-left">Associated BEAST-object</th>
</tr>
</thead>
<tbody>
<tr>
<td class="org-left"><code>run</code></td>
<td class="org-left"><code>beast.core.Runnable</code></td>
</tr>

<tr>
<td class="org-left"><code>distribution</code></td>
<td class="org-left"><code>beast.core.Distribution</code></td>
</tr>

<tr>
<td class="org-left"><code>operator</code></td>
<td class="org-left"><code>beast.core.Operator</code></td>
</tr>

<tr>
<td class="org-left"><code>logger</code></td>
<td class="org-left"><code>beast.core.Logger</code></td>
</tr>

<tr>
<td class="org-left"><code>data</code></td>
<td class="org-left"><code>beast.evolution.alignment.Alignment</code></td>
</tr>

<tr>
<td class="org-left"><code>sequence</code></td>
<td class="org-left"><code>beast.evolution.alignment.Sequence</code></td>
</tr>

<tr>
<td class="org-left"><code>state</code></td>
<td class="org-left"><code>beast.core.State</code></td>
</tr>

<tr>
<td class="org-left"><code>parameter</code></td>
<td class="org-left"><code>beast.core.parameter.RealParameter</code></td>
</tr>

<tr>
<td class="org-left"><code>tree</code></td>
<td class="org-left"><code>beast.evolution.tree.Tree</code></td>
</tr>

<tr>
<td class="org-left"><code>input</code></td>
<td class="org-left">reserved name</td>
</tr>

<tr>
<td class="org-left"><code>map</code></td>
<td class="org-left">macro to simplify <code>spec</code></td>
</tr>

<tr>
<td class="org-left"><code>plate</code></td>
<td class="org-left">for-loop style macro</td>
</tr>
</tbody>
</table>
</div>

<div id="outline-container-orgf064e17" class="outline-4">
<h4 id="orgf064e17">The plate macro</h4>
<div class="outline-text-4" id="text-orgf064e17">
<p>
The XML
</p>

<div class="org-src-container">
<pre class="src src-xml">&lt;plate var="n" range="v1,v2,v3" &gt;
    &lt;parameter id="value.$(n)" a="foo" b="$(n)" /&gt;
&lt;/plate&gt;
</pre>
</div>

<p>
becomes
</p>

<div class="org-src-container">
<pre class="src src-xml">&lt;parameter id="value.v1" a="foo" b="v1" /&gt;
&lt;parameter id="value.v2" a="foo" b="v2" /&gt;
&lt;parameter id="value.v3" a="foo" b="v3" /&gt;
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-orgbb56c34" class="outline-3">
<h3 id="orgbb56c34">Understanding the MCMC application</h3>
<div class="outline-text-3" id="text-orgbb56c34">
<p>
Occasionally, BEAST will dump the state of the chain to a file: for example,
running <code>demo.xml</code> will typically store the state in another file called
<code>demo.xml.state</code>. This state file allows you to resume a sampler at a later date
if you want to continue the chain. The following will demonstrate how to do
this.
</p>

<p>
Grab a BEAST2 XML, we will use <code>demo.xml</code>. We can then run this as follows:
</p>

<div class="org-src-container">
<pre class="src src-sh">java -cp &lt;path/to/beast.jar&gt; beast.app.beastapp.BeastMain -seed 1 demo.xml
</pre>
</div>

<p>
In addition to the log file, this should also produce a file <code>demo.xml.state</code>.
We can rename this file to whatever we want, for example <code>mv demo.xml.state
burnt.xml.state</code>. Then we can restart the sampler at the point saved in that
file:
</p>

<div class="org-src-container">
<pre class="src src-sh">java -cp &lt;path/to/beast.jar&gt; beast.app.beastapp.BeastMain -seed 1 -resume -statefile burnt.xml.state demo.xml
</pre>
</div>

<p>
The flag <code>-resume</code> tells BEAST to append to the end of the existing log file and
<code>-statefile burnt.xml.state</code> specifies which file to read the state from
(otherwise it will look for one called <code>demo.xml.state</code>).
</p>
</div>

<div id="outline-container-org3261f61" class="outline-4">
<h4 id="org3261f61">Starting tree</h4>
<div class="outline-text-4" id="text-org3261f61">
<p>
Navigating tree space can be difficult so it is useful to be able to start a
chain off on a sensible tree. There is a tab in BEAUti (which is hidden by
default) that allows you to define the starting tree, or a distribution to
sample it from. You can make this tab visible via the <b>View</b> menu. In the
<b>Starting tree</b> tab you can choose a distribution to sample a tree from, a
clustering algorithm to generate a tree, or you can specify the tree manually in
Newick format.
</p>

<p>
When specifying the tree via Newick, the following snippets demonstrate how this
is done for BEAST2.6 and BEAST2.7:
</p>

<div class="org-src-container">
<pre class="src src-xml">&lt;!-- 2.6 --&gt;
&lt;init spec='beast.util.TreeParser' id='NewickTree.t:XYZ26'
initial="@Tree.t:XYZ26" taxa='@XYZ26' IsLabelledNewick="true"
newick="((your,(tree,goes)),here)"/&gt;

&lt;!-- 2.7 --&gt;
&lt;init spec='beast.base.evolution.tree.TreeParser' id='NewickTree.t:XYZ26'
initial="@Tree.t:XYZ26" taxa='@XYZ26' IsLabelledNewick="true"
newick="((your,(tree,goes)),here)"/&gt;
</pre>
</div>
</div>
</div>


<div id="outline-container-org2d940a3" class="outline-4">
<h4 id="org2d940a3">Priors</h4>
<div class="outline-text-4" id="text-org2d940a3">
<p>
The <a href="https://github.com/BEAST2-Dev/BEASTLabs/blob/master/src/beastlabs/math/distributions/ExcludablePrior.java"><code>ExcludablePrior</code></a> class from the <a href="https://github.com/BEAST2-Dev/BEASTLabs">BEASTLabs</a> package provides a way to set a
prior distribution on an individual element of a RealParameter. This can be used
with the <code>ScaleOperator</code> when you need a way to change just some parts of a real
vector.
</p>

<p>
To make use of this class you will need to have BEASTLabs installed. The
following snippet demonstrates how to use it. It considers a length three real
parameter and applies a beta prior to the second element, as indicated by the
<code>xInclude</code> attribute.
</p>

<div class="org-src-container">
<pre class="src src-xml">&lt;map name="excludablePrior" &gt;beast.math.distributions.ExcludablePrior&lt;/map&gt;

&lt;excludablePrior id="PRIOR-ID" name="distribution" x="PARAM-ID-REF" xInclude="false true false"&gt;
  &lt;Beta name="distr"&gt;
    &lt;parameter spec="parameter.RealParameter" estimate="false" name="alpha"&gt;2.0&lt;/parameter&gt;
    &lt;parameter spec="parameter.RealParameter" estimate="false" name="beta"&gt;2.0&lt;/parameter&gt;
  &lt;/Beta&gt;
&lt;/excludablePrior&gt;
</pre>
</div>
</div>
</div>

<div id="outline-container-orgeb54722" class="outline-4">
<h4 id="orgeb54722">Operators</h4>
<div class="outline-text-4" id="text-orgeb54722">
<p>
The <code>ScaleOperator</code> has an input <code>indicator</code> which can be used to select
elements of a <code>RealParameter</code> to scale. Frustratingly, there is not the same
functionality for the random walk operators (or any other operator for that
matter). If you want to set a particular parameter element to a particular value
while allowing the others to change, this might be what you need.
</p>

<p>
To use this, you need to construct a Boolean parameter vector. You could do this
by creating one in the <code>state</code> node as a <code>stateNode</code> and then referencing it in
the operator.
</p>
</div>
</div>

<div id="outline-container-org07acfaa" class="outline-4">
<h4 id="org07acfaa">Tuning operators</h4>
<div class="outline-text-4" id="text-org07acfaa">
<p>
Not all operators will auto-tune. If there are integer parameters you may need
to do some small example runs to get a sensible rate of mixing. Some things
online say you should aim for approximately \(30\%\) acceptance of samples.
</p>
</div>
</div>
</div>
</div>

<div id="outline-container-orge446225" class="outline-2">
<h2 id="orge446225">Simulating data</h2>
<div class="outline-text-2" id="text-orge446225">
</div>
<div id="outline-container-org559ab4e" class="outline-3">
<h3 id="org559ab4e">ReMASTER</h3>
<div class="outline-text-3" id="text-org559ab4e">
<p>
ReMASTER is a re-write of the MASTER simulation program. There is an
extensive list of <a href="https://tgvaughan.github.io/remaster/#id-1-Examples">examples</a> in the documentation.
</p>
</div>

<div id="outline-container-orgf35852f" class="outline-4">
<h4 id="orgf35852f">Punctual reactions</h4>
<div class="outline-text-4" id="text-orgf35852f">
<p>
Punctual reactions are those that are applied to a population as a
whole at a particular time. See the <code>PunctualReaction</code> class for
details.
</p>

<ul class="org-ul">
<li><a href="https://github.com/tgvaughan/remaster/blob/master/examples/BDcontemp.xml">Example XML</a></li>
<li><a href="https://tgvaughan.github.io/remaster/#id-4-Reactions-Model-specification-Birth-death-simulations-Punctual-reactions">Documentation</a></li>
</ul>
</div>
</div>
</div>
</div>

<div id="outline-container-org7fbda54" class="outline-2">
<h2 id="org7fbda54">Tools and useful functions</h2>
<div class="outline-text-2" id="text-org7fbda54">
<ul class="org-ul">
<li>You can run both <code>beast</code> and <code>beauti</code> from the command line as they are both
available in the <code>beast.jar</code>. In the case of <code>beauti</code> it will be something like
<code>java -sp &lt;path/to/beast.jar&gt; beast.app.beauti.Beauti</code>.</li>
<li>Tracer, FigTree, TempEst and TreeAnnotator all comes as a separate jar files.
You can just click on the jar file and it should run the program correctly.
<ul class="org-ul">
<li><a href="./phylogenetics-notes.html#org4109702">My TempEst notes</a></li>
</ul></li>
</ul>
</div>

<div id="outline-container-orgd344950" class="outline-3">
<h3 id="orgd344950">Cogsworth</h3>
<div class="outline-text-3" id="text-orgd344950">
<p>
The <b>cogsworth</b> ANT file is intended to simplify some of the typical tasks
involved in running BEAST2 from the command line. You can get a copy <a href="https://github.com/aezarebski/beasts-castle/blob/main/cogsworth.org">here</a>!
</p>
</div>
</div>

<div id="outline-container-org4f66aff" class="outline-3">
<h3 id="org4f66aff">Cogsworth 2.0: snakemake flavoured Cogsworth</h3>
<div class="outline-text-3" id="text-org4f66aff">
<p>
I have found snakemake to be a simpler pipeline tool than ANT. The
following examples repeat some of the Cogsworth functionality, but
with snakemake rules which are probably a bit more friendly.
</p>
</div>

<div id="outline-container-org651bb55" class="outline-4">
<h4 id="org651bb55">Setting up BEAST</h4>
<div class="outline-text-4" id="text-org651bb55">
<div class="org-src-container">
<pre class="src src-snakemake">rule setuplib_beast:
    shell:
        """
        mkdir -p lib
        wget -O lib/BEAST.v2.7.6.Linux.x86.tgz https://github.com/CompEvol/beast2/releases/download/v2.7.6/BEAST.v2.7.6.Linux.x86.tgz
        tar -xzf lib/BEAST.v2.7.6.Linux.x86.tgz -C lib/
        chmod 750 lib/beast/bin/beast
        chmod 750 lib/beast/bin/beauti
        chmod 750 lib/beast/jre/bin/java
        """
</pre>
</div>
</div>
</div>

<div id="outline-container-org1f96516" class="outline-4">
<h4 id="org1f96516">Setting up Tracer</h4>
<div class="outline-text-4" id="text-org1f96516">
<div class="org-src-container">
<pre class="src src-snakemake">rule setuplib_tracer:
    shell:
        """
        mkdir -p lib
        wget -O lib/Tracer_v1.7.2.tgz https://github.com/beast-dev/tracer/releases/download/v1.7.2/Tracer_v1.7.2.tgz
        tar -xzf lib/Tracer_v1.7.2.tgz -C lib/
        """
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-org932135b" class="outline-3">
<h3 id="org932135b">Web beautify</h3>
<div class="outline-text-3" id="text-org932135b">
<p>
Having a linter for XML is helpful. The tool js-beautify seems decent: <a href="https://beautifier.io/">LINK</a> and
<a href="https://github.com/beautify-web/js-beautify">GitHub</a>. You can configure this if you do not like the default style it uses.
Just add the following as <code>.jsbeautifyrc</code> in the root of your project.
</p>

<div class="org-src-container">
<pre class="src src-json">{
    "indent_size": 4,
    "preserve_newlines": true,
    "wrap_attributes": "force-aligned"
}
</pre>
</div>

<p>
The following function handles calling the beautifier from within R
</p>

<div class="org-src-container">
<pre class="src src-R">#' Run js-beautify on the given file.
#'
#' @param x is the path to the XML file.
#'
beautify_xml &lt;- function(x) {
  system2(command = "html-beautify", args = c("-f", x, "-r"))
}
</pre>
</div>
</div>
</div>

<div id="outline-container-sec:read-beast2-log" class="outline-3">
<h3 id="sec:read-beast2-log">Read BEAST2 log file</h3>
<div class="outline-text-3" id="text-sec:read-beast2-log">
<p>
The following function reads a log file into a suitable dataframe.
</p>

<div class="org-src-container">
<pre class="src src-R">#' Read a BEAST2 log file into a data frame.
#'
#' @param filename is the path to the log file.
#' @param burn is the number to remove from the start.
#' @param take_last is the number to take from the end.
#'
#' @return data frame containing the samples.
#'
read_beast2_log &lt;- function(filename, burn=0, take_last=NA) {
  y &lt;- read.csv(filename, sep = "\t", comment.char = "#")
  if (is.na(take_last) &amp;&amp; burn &gt;= 0) {
    return(tail(y, nrow(y) - burn))
  } else if (!is.na(take_last) &amp;&amp; burn == 0) {
    return(tail(y, take_last))
  } else {
    stop("Unsupported arguments given to read_beast2_log.")
  }
}
</pre>
</div>
</div>
</div>

<div id="outline-container-org58dfa98" class="outline-3">
<h3 id="org58dfa98">Running BEAST2 XML files with Python</h3>
<div class="outline-text-3" id="text-org58dfa98">
<p>
The <code>run_beast2_simulations_parallel</code> takes a list of XML files and a
number of workers and then distributes the running of these files
across the workers. The <code>subprocess</code> package is used to run BEAST2 and
we use the <code>ThreadPoolExecutor</code> (and <code>as_completed</code>) from
<code>concurrent.futures</code> to manage the workers.
</p>

<p>
This is particularly useful if you are running a whole heap of MCMC
files as part of a simulation study, or trying to generate a lot of
datasets with something like remaster.
</p>

<div class="org-src-container">
<pre class="src src-python">def run_beast2_simulations_parallel(simulation_xml_list, num_jobs):
    def run_beast2(simulation_xml):
        """
        Run a BEAST2 simulation using the provided XML file.

        If the simulation does not finish within 5 minutes, it is
        considered to have timed out.

        $ ./lib/beast/bin/beast -seed 1 -overwrite &lt;simulation_xml&gt;
        """
        print(f"Running simulation: {simulation_xml}")
        beast_executable_mac = "/Applications/BEAST 2.7.6/bin/beast"
        beast_executable_linux = "./lib/beast/bin/beast"
        if os.path.exists(beast_executable_mac):
            beast_executable = beast_executable_mac
        elif os.path.exists(beast_executable_linux):
            beast_executable = beast_executable_linux
        else:
            raise Exception("BEAST2 executable not found.")
        command = [beast_executable, "-seed", "1", "-overwrite", simulation_xml]
        try:
            result = subprocess.run(
                command,
                check=True,
                capture_output=True,
                text=True,
                timeout=300,
            )
            return result.stdout
        except subprocess.TimeoutExpired:
            return f"BEAST2 simulation for {simulation_xml} timed out."
        except subprocess.CalledProcessError as e:
            return f"Error occurred while running BEAST2 simulation for {simulation_xml}: {e.stderr}"

    with ThreadPoolExecutor(max_workers=num_jobs) as executor:
        future_to_xml = {
            executor.submit(run_beast2, xml): xml for xml in simulation_xml_list
        }
        for future in as_completed(future_to_xml):
            xml = future_to_xml[future]
            try:
                data = future.result()
                print(f"Simulation completed for {xml}: {data}")
            except Exception as exc:
                print(f"Simulation generated an exception for {xml}: {exc}")
</pre>
</div>

<p>
Note that this assumes that the executable for BEAST2 is in one of two
standard locations. This should be easy to fix if you have a different
path.
</p>
</div>
</div>

<div id="outline-container-orgb0431cf" class="outline-3">
<h3 id="orgb0431cf">Pretty printing Tracer output</h3>
<div class="outline-text-3" id="text-orgb0431cf">
<p>
The following little script can be used to translate a summary produced by
Tracer into something that is easier to use in documents.
</p>

<div class="org-src-container">
<pre class="src src-R">#!/usr/bin/env Rscript
# -*- mode:ess-mode; -*-
#'
#' tracer-table
#' ============
#'
#' Tool to convert tracer summary to an org-mode/markdown style table.
#'
#' Usage
#' -----
#'
#' ./tracer-table --input &lt;demo.txt&gt; --output demo.org
#'

suppressPackageStartupMessages(library(argparse))

parser &lt;- ArgumentParser()

parser$add_argument(
         "-v",
         "--verbose",
         action = "store_true",
         default = FALSE,
         help = "Verbose output"
       )
parser$add_argument(
         "-i",
         "--input",
         type = "character",
         help = "Input file"
       )
parser$add_argument(
         "-o",
         "--output",
         type = "character",
         help = "Output file"
       )

args &lt;- parser$parse_args()


main &lt;- function(args) {
  foo &lt;- readLines(args$input) |&gt;
    gsub(pattern = "\t", replacement = " | ") |&gt;
    gsub(pattern = "^", replacement = "| ") |&gt;
    gsub(pattern = "$", replacement = " |")
  tbl_header &lt;- head(foo, 1)
  tbl_hline &lt;- tbl_header |&gt;
    gsub(pattern = "[_\ a-zA-Z]", replacement = "-") |&gt;
    gsub(pattern = "-\\|-", replacement = "-+-")
  tbl_body &lt;- tail(foo, -1)
  bar &lt;- c(tbl_header, tbl_hline, tbl_body)
  writeLines(text = bar, con = args$output)
}

if (!interactive()) {
  args &lt;- parser$parse_args()
  main(args)
}
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-org2697552" class="outline-2">
<h2 id="org2697552">Examples</h2>
<div class="outline-text-2" id="text-org2697552">
<ul class="org-ul">
<li><a href="#orgbd9ccea">BEAST2 as an MCMC engine: part I</a></li>
<li><a href="#org2abb9f1">BEAST2 as an MCMC engine: part II</a></li>
<li><a href="#org7ecb81e">Birth-death simulation: Example I</a></li>
<li><a href="#orga2df028">Birth-death simulation: Example II</a></li>
</ul>
</div>

<div id="outline-container-orgbd9ccea" class="outline-3">
<h3 id="orgbd9ccea">BEAST2 as an MCMC engine: part I</h3>
<div class="outline-text-3" id="text-orgbd9ccea">
<p>
This example looks at how to use BEAST2 to simulate from a standard normal
distribution using MCMC. The XML for this has the following structure
</p>

<div class="org-src-container">
<pre class="src src-xml">&lt;?xml version="1.0" encoding="UTF-8" standalone="no"?&gt;
&lt;beast version="2.0"&gt;

  &lt;&lt;targetDistribution&gt;&gt;

  &lt;&lt;mcmc&gt;&gt;

&lt;/beast&gt;
</pre>
</div>

<p>
The <a href="#org6c4ac0f"><code>targetDistribution</code></a> specifies the distribution we will sample from and the <a href="#org1646c6f">mcmc</a> specifies how details of the MCMC algorithm.
</p>
</div>

<div id="outline-container-org6c4ac0f" class="outline-4">
<h4 id="org6c4ac0f">The target distribution</h4>
<div class="outline-text-4" id="text-org6c4ac0f">
<p>
We use a <code>Prior</code> to represent the distribution because this makes it easy to
evaluate the distribution at a particular value.
</p>

<div class="org-src-container">
<pre class="src src-xml" id="org973f57f">&lt;!-- targetDistribution --&gt;
&lt;distribution id="posterior" spec="beast.base.inference.distribution.Prior" x="@x"&gt;
  &lt;distr spec="beast.base.inference.distribution.Normal" mean="0.0" sigma="1.0" /&gt;
&lt;/distribution&gt;
</pre>
</div>
</div>
</div>

<div id="outline-container-org1646c6f" class="outline-4">
<h4 id="org1646c6f">MCMC</h4>
<div class="outline-text-4" id="text-org1646c6f">
<p>
We use a <code>run</code> tag to specify the MCMC. This requires a <code>state</code> object to
describe the parameters we are sampling, the <code>distribution</code> over the state, and
an <code>operator</code> and <code>logger</code> to perturb the state and record the results.
</p>

<div class="org-src-container">
<pre class="src src-xml" id="orge7511e4">&lt;!-- mcmc --&gt;
&lt;run chainLength="10000000" id="mcmc" preBurnin="100" spec="beast.base.inference.MCMC"&gt;

    &lt;state id="state" storeEvery="1000"&gt;
        &lt;parameter estimate="true" id="x" name="stateNode" value="1.0" /&gt;
    &lt;/state&gt;

    &lt;distribution idref="posterior" /&gt;

    &lt;operator spec="beast.base.inference.operator.RealRandomWalkOperator"
              windowSize="0.1"
              useGaussian="true"
              weight="1.0"
              parameter="@x" /&gt;

    &lt;logger id="screenlog" logEvery="1000000"&gt;
        &lt;log idref="x" /&gt;
    &lt;/logger&gt;

    &lt;logger fileName="beast.csv"
            id="tracelog"
            logEvery="10000"
            model="@posterior"&gt;
        &lt;log idref="posterior" /&gt;
        &lt;log idref="x" /&gt;
    &lt;/logger&gt;

&lt;/run&gt;
</pre>
</div>
</div>
</div>

<div id="outline-container-org97d7c9f" class="outline-4">
<h4 id="org97d7c9f">Running the sampler</h4>
<div class="outline-text-4" id="text-org97d7c9f">
<p>
Here is the command to actually run this program.
</p>

<div class="org-src-container">
<pre class="src src-sh">java -jar beast.jar -seed 1 -overwrite example-01.xml
</pre>
</div>

<p>
I have used the <code>-overwrite</code> flag because otherwise it will prompt you each time
unless you change the output file name.
</p>
</div>
</div>

<div id="outline-container-org12c78ce" class="outline-4">
<h4 id="org12c78ce">Visualising the results</h4>
<div class="outline-text-4" id="text-org12c78ce">
<p>
Here is a visualisation of the resulting samples along with the distribution we
were sampling from.
</p>


<div id="org6f269f0" class="figure">
<p><img src="../images/beast-demo.png" alt="beast-demo.png" width="800px" />
</p>
</div>

<div class="org-src-container">
<pre class="src src-R">library(ggplot2)
library(cowplot)

beast_samples &lt;- read.table("beast.csv",
  sep = "\t",
  comment.char = "#",
  header = TRUE
)

g1 &lt;- ggplot() +
  geom_line(
    data = beast_samples,
    mapping = aes(x = Sample, y = x)
  ) +
  labs(x = "Sample", y = "X") +
  theme_classic()

g2 &lt;- ggplot() +
  geom_histogram(
    data = beast_samples,
    mapping = aes(x = x, y = ..density..),
    bins = 10
  ) +
  stat_function(
    fun = \(x) dnorm(x),
    geom = "line"
  ) +
  labs(x = "X", y = "Density") +
  theme_classic()

g &lt;- plot_grid(g1, g2)

ggsave("beast-demo.png",
       g,
       height = 10.5,
       width = 14.8,
       units = "cm")
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-org2abb9f1" class="outline-3">
<h3 id="org2abb9f1">BEAST2 as an MCMC engine: part II</h3>
<div class="outline-text-3" id="text-org2abb9f1">
<p>
This example looks at estimating the probability of heads in coin flips. We
start with at beta prior distribution and take values of the number of observed
heads and tails from the command line.
</p>

<div class="org-src-container">
<pre class="src src-xml">&lt;?xml version="1.0" encoding="UTF-8" standalone="no"?&gt;
&lt;beast version="2.0"&gt;

    &lt;distribution id="posterior" spec="beast.core.util.CompoundDistribution" &gt;
        &lt;distribution id="prior" spec="beast.math.distributions.Prior" x="@p"&gt;
            &lt;distr spec="beast.math.distributions.Beta" alpha="0.5" beta="0.5" /&gt;
        &lt;/distribution&gt;

        &lt;distribution id="likelihood" spec="beast.math.distributions.Prior" x="@p"&gt;
            &lt;distr spec="beast.math.distributions.Beta" alpha="$(heads)" beta="$(tails)" /&gt;
        &lt;/distribution&gt;
    &lt;/distribution&gt;

    &lt;run chainLength="10000000" id="mcmc" preBurnin="100" spec="beast.core.MCMC"&gt;

        &lt;state id="state" storeEvery="1000"&gt;
            &lt;parameter estimate="true" id="p" name="stateNode" value="0.5" /&gt;
        &lt;/state&gt;

        &lt;distribution idref="posterior" /&gt;

        &lt;operator spec="beast.evolution.operators.RealRandomWalkOperator" windowSize="0.1" useGaussian="true" weight="1.0" parameter="@p" /&gt;

        &lt;logger id="screenlog" logEvery="1000000"&gt;
            &lt;log idref="p" /&gt;
        &lt;/logger&gt;

        &lt;logger fileName="beast-02.csv" id="tracelog" logEvery="10000" model="@posterior"&gt;
            &lt;log idref="posterior" /&gt;
            &lt;log idref="p" /&gt;
        &lt;/logger&gt;

    &lt;/run&gt;

&lt;/beast&gt;
</pre>
</div>
</div>

<div id="outline-container-org78ed253" class="outline-4">
<h4 id="org78ed253">Running the sampler</h4>
<div class="outline-text-4" id="text-org78ed253">
<p>
Here is the command to actually run this program.
</p>

<div class="org-src-container">
<pre class="src src-sh">java -jar beast.jar -seed 1 -overwrite -D "heads=3,tails=7" example-02.xml
</pre>
</div>

<p>
The <code>-D</code> flag has been used to specify how many times the coin came up heads and
tails.
</p>
</div>
</div>

<div id="outline-container-org0045682" class="outline-4">
<h4 id="org0045682">Visualising the results</h4>
<div class="outline-text-4" id="text-org0045682">
<p>
Here is a visualisation of the resulting samples along with the distribution we
were sampling from.
</p>


<div id="org0b0d80b" class="figure">
<p><img src="../images/beast-02.png" alt="beast-02.png" width="800px" />
</p>
</div>

<div class="org-src-container">
<pre class="src src-R">library(ggplot2)
library(cowplot)

beast_samples &lt;- read.table("beast-02.csv",
  sep = "\t",
  comment.char = "#",
  header = TRUE
)

g1 &lt;- ggplot() +
  geom_line(
    data = beast_samples,
    mapping = aes(x = Sample, y = p)
  ) +
  labs(x = "Sample", y = "p") +
  theme_classic()

g2 &lt;- ggplot() +
  geom_histogram(
    data = beast_samples,
    mapping = aes(x = p, y = ..density..),
    bins = 15
  ) +
  stat_function(
    fun = \(x) dbeta(x, 0.5 + 3, 0.5 + 7),
    geom = "line"
  ) +
  labs(x = "X", y = "Density") +
  theme_classic()

g &lt;- plot_grid(g1, g2)

ggsave("beast-02.png",
       g,
       height = 10.5,
       width = 14.8,
       units = "cm")
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-org7ecb81e" class="outline-3">
<h3 id="org7ecb81e">Birth-death simulation: Example I</h3>
<div class="outline-text-3" id="text-org7ecb81e">
<p>
In this example we will simulate sequences on a realisation of the birth-death
process and then attempt to estimate the birth rate from these sequences. We
will use <code>ape</code> to simulate the birth-death process and <code>phangorn</code> to simulate the
sequences. To ensure a correct XML specification we will use <code>beauti</code> to generate
the XML for the analysis. Finally, we will use <code>tracer</code> to check the results.
</p>
</div>

<div id="outline-container-org20f045a" class="outline-4">
<h4 id="org20f045a">Simulating the tree</h4>
<div class="outline-text-4" id="text-org20f045a">
<p>
We will use years as our unit of time. Assuming that an individual is infectious
for 7 days (\(1/52\) of a year) on average, and that \(1\%\) of infections are
sequenced we end up with the following equations for the rates: \(\mu + \psi =
52\) and \(\psi / (\psi + \mu) = 1 / 100\). Solving these give us \(\mu =
5148/100\) and \(\psi = 52/100\). If we have an \(\lambda / (\mu + \psi) = R_{0}
= 2\) then \(\lambda = 104\).
</p>

<div class="org-src-container">
<pre class="src src-R" id="org40cf12a">birth_rate &lt;- 104
death_rate &lt;- 5148 / 100
sampling_rate &lt;- 52 / 100
sim_duration &lt;- 0.15 # year
</pre>
</div>

<p>
The <code>rlineage</code> function returns a realisation of the birth-death process as a
<code>phylo</code> object.
</p>

<div class="org-src-container">
<pre class="src src-R" id="org8202bbc">net_removal_rate &lt;- death_rate + sampling_rate
sim_tree &lt;- rlineage(birth_rate,
                     net_removal_rate,
                     Tmax = sim_duration)
</pre>
</div>

<p>
We include the tip dates as part of the tip labels because it makes it much
easier to read the data into Beauti if you can parse this information from the
top labels.
</p>

<div class="org-src-container">
<pre class="src src-R" id="org63504fb">num_tips &lt;- length(sim_tree$tip.label)
tip_fwd_times &lt;- node.depth.edgelength(sim_tree)[1:num_tips]
sim_tree$tip.label &lt;- sprintf("%s_%f",
                              sim_tree$tip.label,
                              2021 + tip_fwd_times)
</pre>
</div>

<p>
The reconstructed tree is the tree that remains when we sample the extinct tips
(ie those that are not extant). The probability that a tip is sequenced is
\(\psi / (\mu + \psi\)\).
</p>

<div class="org-src-container">
<pre class="src src-R" id="org83a2de4">extant_mask &lt;- tip_fwd_times &gt;= sim_duration
num_removed &lt;- sum(!extant_mask)
num_sampled &lt;- rbinom(n = 1,
                      size = num_removed,
                      prob = sampling_rate / net_removal_rate)
sampled_tips &lt;- sample(x = sim_tree$tip.label[!extant_mask],
                       size = num_sampled)
reconstructed_tree &lt;- keep.tip(phy = sim_tree,
                               tip = sampled_tips)
</pre>
</div>

<p>
We can combine all of this into a function
</p>

<div class="org-src-container">
<pre class="src src-R" id="org71ef7b8">rand_reconstructed_tree &lt;- function(birth_rate,
                                    death_rate,
                                    sampling_rate,
                                    duration) {
  net_removal_rate &lt;- death_rate + sampling_rate
  sim_tree &lt;- rlineage(birth_rate,
                       net_removal_rate,
                       Tmax = sim_duration)

  num_tips &lt;- length(sim_tree$tip.label)
  tip_fwd_times &lt;- node.depth.edgelength(sim_tree)[1:num_tips]
  sim_tree$tip.label &lt;- sprintf("%s_%f",
                                sim_tree$tip.label,
                                2021 + tip_fwd_times)

  extant_mask &lt;- tip_fwd_times &gt;= sim_duration
  num_removed &lt;- sum(!extant_mask)
  num_sampled &lt;- rbinom(n = 1,
                        size = num_removed,
                        prob = sampling_rate / net_removal_rate)
  sampled_tips &lt;- sample(x = sim_tree$tip.label[!extant_mask],
                         size = num_sampled)
  reconstructed_tree &lt;- keep.tip(phy = sim_tree,
                                 tip = sampled_tips)

  return(reconstructed_tree)
}
</pre>
</div>

<p>
We want a simulation that has a reasonable number of sequences in it, so we
generate random trees until we have one with at least \(20\) tips.
</p>

<div class="org-src-container">
<pre class="src src-R">iter_count &lt;- 0
num_seqs &lt;- 0
while (iter_count &lt; 100 &amp;&amp; num_seqs &lt; 20) {
  reconstructed_tree &lt;- rand_reconstructed_tree(birth_rate,
                                                death_rate,
                                                sampling_rate,
                                                sim_duration)
  iter_count &lt;- iter_count + 1
  num_seqs &lt;- length(reconstructed_tree$tip.label)
  print(iter_count)
}
</pre>
</div>

<p>
This tree should be saved for later use.
</p>

<div class="org-src-container">
<pre class="src src-R">write.tree(phy = reconstructed_tree,
           file = "reconstructed-tree.newick")
</pre>
</div>
</div>
</div>

<div id="outline-container-orgf061294" class="outline-4">
<h4 id="orgf061294">Simulating the sequences</h4>
<div class="outline-text-4" id="text-orgf061294">
<p>
Then we use the <code>simSeq</code> function to simulate a set of genomes for the leaves of
this tree. By default this function uses the JC model for the sequence
simulation with a rate of 1 (using the branch lengths as the units of time). RNA
viruses have a mutation rate of \(\approx 10^{-3}\) substitutions per site per
year. We are already working in units of years so we can use this as our rate.
The length of the sequence we will use is 2000 which is similar to the HA of
influenza.
</p>

<div class="org-src-container">
<pre class="src src-R">sim_seqs &lt;- simSeq(x = reconstructed_tree,
                   rate = 1e-3,
                   l = 2000,
                   type = "DNA")
</pre>
</div>

<p>
Then we write this to a file so we can read it into Beauti.
</p>

<div class="org-src-container">
<pre class="src src-R">write.phyDat(x = sim_seqs,
             file = "sequences.fasta",
             format = "fasta")
</pre>
</div>
</div>
</div>

<div id="outline-container-org3b8c12f" class="outline-4">
<h4 id="org3b8c12f">Constructing the XML</h4>
<div class="outline-text-4" id="text-org3b8c12f">
<div class="org-src-container">
<pre class="src src-sh">java -cp ~/Documents/beast2/build/dist/beast.jar beast.app.beauti.Beauti
</pre>
</div>


<ol class="org-ol">
<li>Open Beauti and load in the alignment produced by the simulation above.</li>
<li>Open the <b>Tip Dates</b> tab and auto-configure to get the tip dates.
<ul class="org-ul">
<li>Since the sequences have been simulated with JC model with rate 1 nothing
special is needed for the <b>Site Model</b> and <b>Clock Model</b> tabs.</li>
</ul></li>
<li>Open the <b>Priors</b> tab and select the <b>Birth Death Skyline Serial</b> prior
<ul class="org-ul">
<li>The reproduction number has dimension 10 by default, we used constant rates
so this should be reduced to 1.</li>
<li>Go over the parameters and make sure that for each one the prior looks
sensible. To make things a bit easier, set informative priors for each
parameter.</li>
</ul></li>
<li>Save the analysis as an XML.</li>
</ol>
</div>
</div>

<div id="outline-container-org625f754" class="outline-4">
<h4 id="org625f754">Running the sampler</h4>
<div class="outline-text-4" id="text-org625f754">
<p>
Running the sampler for \(10^{6}\) iterations does not take very long and allows
us to see which variables are getting stuck. There are some suggested
adjustments to the operators which should be done before increasing the sample
size to \(10^{7}\).
</p>
</div>
</div>

<div id="outline-container-org572676e" class="outline-4">
<h4 id="org572676e">Analysing the posterior samples</h4>
<div class="outline-text-4" id="text-org572676e">
<p>
Tracer can be used for preliminary investigation of the posterior samples as
shown in the following table (generated using <a href="#orgb0431cf"><code>tracer-table</code></a>).
</p>

<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">


<colgroup>
<col  class="org-left" />

<col  class="org-right" />

<col  class="org-left" />

<col  class="org-right" />
</colgroup>
<thead>
<tr>
<th scope="col" class="org-left">Summary Statistic</th>
<th scope="col" class="org-right">Mean</th>
<th scope="col" class="org-left">95% HPD Interval</th>
<th scope="col" class="org-right">Effective Sample Size</th>
</tr>
</thead>
<tbody>
<tr>
<td class="org-left">clock rate</td>
<td class="org-right">7.641E-4</td>
<td class="org-left">[3.9092E-4, 1.171E-3]</td>
<td class="org-right">496.5</td>
</tr>

<tr>
<td class="org-left">origin time</td>
<td class="org-right">0.1509</td>
<td class="org-left">[0.1146, 0.1935]</td>
<td class="org-right">254.7</td>
</tr>

<tr>
<td class="org-left">become uninfectious rate</td>
<td class="org-right">51.9713</td>
<td class="org-left">[50.0723, 53.9649]</td>
<td class="org-right">4406.5</td>
</tr>

<tr>
<td class="org-left">reproductive number</td>
<td class="org-right">2.0901</td>
<td class="org-left">[1.8752, 2.3189]</td>
<td class="org-right">581.7</td>
</tr>

<tr>
<td class="org-left">sampling proportion</td>
<td class="org-right">0.0204</td>
<td class="org-left">[1.9394E-3, 0.0435]</td>
<td class="org-right">116.5</td>
</tr>
</tbody>
</table>

<p>
To display the prior and posteriors on the same figure we need to resort to
<code>ggplot2</code> though. There appear to be some identifiability issues (which is to be
expected), but the estimates of the origin and basic reproduction number are
decent. The figure shows the prior distributions (solid lines), the marginal
posterior distribution (histogram) and the true values used in the simulation
(dashed red line).
</p>


<div id="orgd1d2a11" class="figure">
<p><img src="../misc/beast2/bd-sim-ex-1/combined-results.png" alt="combined-results.png" width="800px" />
</p>
</div>

<div class="org-src-container">
<pre class="src src-R">library(dplyr)
library(ggplot2)
library(magrittr)
library(purrr)
library(cowplot)

&lt;&lt;define-parameters&gt;&gt;

mcmc_samples &lt;- read.csv(
  "sequences.log",
  comment.char = "#",
  sep = "\t"
) |&gt;
  rename(
    clock_rate = clockRate,
    origin = origin_BDSKY_Serial,
    net_removal_rate = becomeUninfectiousRate_BDSKY_Serial,
    r_naught = reproductiveNumber_BDSKY_Serial,
    sampling_proportion = samplingProportion_BDSKY_Serial
  ) |&gt;
  select(
    clock_rate,
    origin,
    net_removal_rate,
    r_naught,
    sampling_proportion
  ) |&gt;
  tail(-1000)


g1 &lt;- ggplot() +
  geom_histogram(
    data = mcmc_samples,
    mapping = aes(x = r_naught, y = ..density..),
    bins = 40
  ) +
  stat_function(
    fun = function(x) dlnorm(x, meanlog = 0.0, sdlog = 1.0),
    geom = "line"
  ) +
  geom_vline(
    xintercept = birth_rate / (death_rate + sampling_rate),
    linetype = "dashed",
    colour = "red"
  ) +
  xlim(c(0.1, 3)) +
  labs(x = "R-naught", y = "Density") +
  theme_classic()

g2 &lt;- ggplot() +
  geom_histogram(
    data = mcmc_samples,
    mapping = aes(
      x = net_removal_rate,
      y = ..density..
    )
  ) +
  stat_function(
    fun = function(x) dnorm(x, mean = 52, sd = 1.0),
    geom = "line"
  ) +
  geom_vline(
    xintercept = death_rate + sampling_rate,
    linetype = "dashed",
    colour = "red"
  ) +
  xlim(c(48, 56)) +
  labs(x = "Net removal rate", y = "Density") +
  theme_classic()

g3 &lt;- ggplot() +
  geom_histogram(
    data = mcmc_samples,
    mapping = aes(x = sampling_proportion, y = ..density..),
    bins = 20
  ) +
  stat_function(
    fun = function(x) dbeta(x, shape1 = 2.0, shape2 = 97.0),
    geom = "line"
  ) +
  geom_vline(
    xintercept = sampling_rate / (death_rate + sampling_rate),
    linetype = "dashed",
    colour = "red"
  ) +
  xlim(c(0, 0.1)) +
  labs(x = "Sampling proportion", y = "Density") +
  theme_classic()

g4 &lt;- ggplot() +
  geom_histogram(
    data = mcmc_samples,
    mapping = aes(x = origin, y = ..density..),
    bins = 20
  ) +
  stat_function(
    fun = function(x) dlnorm(x, meanlog = log(0.1), sdlog = 1.25),
    geom = "line"
  ) +
  geom_vline(
    xintercept = sim_duration,
    linetype = "dashed",
    colour = "red"
  ) +
  xlim(c(0, 0.3)) +
  labs(x = "Origin", y = "Density") +
  theme_classic()

g5 &lt;- ggplot() +
  geom_histogram(
    data = mcmc_samples,
    mapping = aes(x = clock_rate, y = ..density..),
    bins = 40
  ) +
  stat_function(
    fun = function(x) dgamma(x, 10, 1e4),
    geom = "line"
  ) +
  geom_vline(
    xintercept = 1e-3,
    linetype = "dashed",
    colour = "red"
  ) +
  xlim(c(1e-5, 3e-3)) +
  labs(x = "Clock rate", y = "Density") +
  theme_classic()

g_comb &lt;- plot_grid(
  plot_grid(g1, g2, g3, nrow = 1, labels = LETTERS[1:3]),
  plot_grid(g4, g5, nrow = 1, labels = LETTERS[4:5]),
  nrow = 2
)

ggsave(
  filename = "combined-results.png",
  plot = g_comb,
  height = 14.8, width = 21.0,
  units = "cm"
)
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-orga2df028" class="outline-3">
<h3 id="orga2df028">Birth-death simulation: Example II</h3>
<div class="outline-text-3" id="text-orga2df028">
<p>
In this example we will do pretty much the same thing as above, but will
simulate a second smaller epidemic in a different location where the sampling
proportion is two-thirds of the size. The goal it to analyse the two datasets
simultaneously to get tighter estimates on the common parameters but also get
the individual parameters.
</p>

<div class="org-src-container">
<pre class="src src-R">&lt;&lt;import-libraries&gt;&gt;

birth_rate &lt;- 104
death_rate &lt;- 5148 / 100
sampling_rate_1 &lt;- 52 / 100
sampling_rate_2 &lt;- 0.66 * 52 / 100
sim_duration &lt;- 0.15 # year

&lt;&lt;define-random-sampler&gt;&gt;

iter_count &lt;- 0
num_seqs &lt;- 0
while (iter_count &lt; 100 &amp;&amp; num_seqs &lt; 20) {
  reconstructed_tree_1 &lt;- rand_reconstructed_tree(
    birth_rate,
    death_rate,
    sampling_rate_1,
    sim_duration)
  iter_count &lt;- iter_count + 1
  num_seqs &lt;- length(reconstructed_tree_1$tip.label)
}

write.tree(phy = reconstructed_tree_1,
           file = "reconstructed-tree-1.newick")

sim_seqs_1 &lt;- simSeq(x = reconstructed_tree_1,
                     rate = 1e-3,
                     l = 2000,
                     type = "DNA")

write.phyDat(x = sim_seqs_1,
             file = "sequences-1.fasta",
             format = "fasta")

iter_count &lt;- 0
num_seqs &lt;- 0
while (iter_count &lt; 100 &amp;&amp; num_seqs &lt; (0.66 * 20)) {
  reconstructed_tree_2 &lt;- rand_reconstructed_tree(
    birth_rate,
    death_rate,
    sampling_rate_2,
    sim_duration)
  iter_count &lt;- iter_count + 1
  num_seqs &lt;- length(reconstructed_tree_2$tip.label)
}

write.tree(phy = reconstructed_tree_2,
           file = "reconstructed-tree-2.newick")

sim_seqs_2 &lt;- simSeq(x = reconstructed_tree_2,
                     rate = 1e-3,
                     l = 2000,
                     type = "DNA")

write.phyDat(x = sim_seqs_2,
             file = "sequences-2.fasta",
             format = "fasta")
</pre>
</div>

<p>
The resulting sequences can be given to Beauti and we can ask it to link the
genetic models. The generated XML has different tree priors but we want to link
the reproduction number and the becoming uninfectious rate and clock rate across
the two datasets so a bit of care needs to be taken to ensure that there is only
a single parameter for each of these things defined and that it is reused and
only has a single prior.
</p>

<p>
To ensure a fair comparison with the previous analysis we can copy the prior
specifications over. Although in doing this some minor tweaks need to be made to
ensure that every ID is unique.
</p>

<p>
Running the analysis and looking at the estimates in Tracer we get the following
table of estimates and HPDs which are very similar to those from the previous
analysis which is perhaps not that surprising given we haven't added all that
much additional data and the estimates were already tight. Note that the ratio
of sampling proportions is close to the true ratio.
</p>

<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">


<colgroup>
<col  class="org-left" />

<col  class="org-right" />

<col  class="org-left" />

<col  class="org-right" />
</colgroup>
<thead>
<tr>
<th scope="col" class="org-left">Summary Statistic</th>
<th scope="col" class="org-right">Mean</th>
<th scope="col" class="org-left">95% HPD Interval</th>
<th scope="col" class="org-right">Effective Sample Size (ESS)</th>
</tr>
</thead>
<tbody>
<tr>
<td class="org-left">clock rate</td>
<td class="org-right">9.1163E-4</td>
<td class="org-left">[5.2452E-4, 1.3203E-3]</td>
<td class="org-right">845.3</td>
</tr>

<tr>
<td class="org-left">origin time 1</td>
<td class="org-right">0.1455</td>
<td class="org-left">[0.1139, 0.1847]</td>
<td class="org-right">378.6</td>
</tr>

<tr>
<td class="org-left">origin time 2</td>
<td class="org-right">0.1417</td>
<td class="org-left">[0.1024, 0.1872]</td>
<td class="org-right">516</td>
</tr>

<tr>
<td class="org-left">become uninfectious rate</td>
<td class="org-right">52.0074</td>
<td class="org-left">[50.0846, 54.0113]</td>
<td class="org-right">8307.3</td>
</tr>

<tr>
<td class="org-left">reproductive number</td>
<td class="org-right">2.1148</td>
<td class="org-left">[1.9335, 2.3078]</td>
<td class="org-right">836.8</td>
</tr>

<tr>
<td class="org-left">sampling proportion 1</td>
<td class="org-right">0.024</td>
<td class="org-left">[4.511E-3, 0.0504]</td>
<td class="org-right">92</td>
</tr>

<tr>
<td class="org-left">sampling proportion 2</td>
<td class="org-right">0.0146</td>
<td class="org-left">[1.0711E-3, 0.033]</td>
<td class="org-right">437.3</td>
</tr>
</tbody>
</table>
</div>
</div>

<div id="outline-container-orgb779836" class="outline-3">
<h3 id="orgb779836">Simulating sequences</h3>
<div class="outline-text-3" id="text-orgb779836">
<p>
The following XML will simulate sequences for the taxons in the data tag based
on a tree read from a Newick file (this requires feast to be installed.) The
results are written to an XML because they take the form of a data block that
you might use in subsequent analyses.
</p>

<div class="org-src-container">
<pre class="src src-xml">&lt;beast version='2.0' namespace='beast.base.evolution.alignment:beast.pkgmgmt:beast.base.core:beast.base.inference:beast.base.evolution.tree.coalescent:beast.pkgmgmt:beast.base.core:beast.base.inference.util:beast.evolution.nuc:beast.base.evolution.operator:beast.base.inference.operator:beast.evolution.sitemodel:beast.evolution.substitutionmodel:beast.base.evolution.likelihood'&gt;


    &lt;data id="alignment" dataType="nucleotide"&gt;
        &lt;sequence taxon="human"&gt;?&lt;/sequence&gt;
        &lt;sequence taxon="chimp"&gt;?&lt;/sequence&gt;
        &lt;sequence taxon="bonobo"&gt;?&lt;/sequence&gt;
        &lt;sequence taxon="gorilla"&gt;?&lt;/sequence&gt;
        &lt;sequence taxon="orangutan"&gt;?&lt;/sequence&gt;
        &lt;sequence taxon="siamang"&gt;?&lt;/sequence&gt;
    &lt;/data&gt;


    &lt;tree spec='feast.fileio.TreeFromNewickFile' fileName="demo-tree.newick" IsLabelledNewick="true" adjustTipHeights="false" id='tree' /&gt;


    &lt;run spec="beast.app.seqgen.SequenceSimulator" id="seqgen" data='@alignment' tree='@tree' sequencelength='100' outputFileName="demo-seqgen-output.xml"&gt;
        &lt;siteModel spec='SiteModel'&gt;
            &lt;substModel spec='JukesCantor' /&gt;
        &lt;/siteModel&gt;

        &lt;branchRateModel id="StrictClock" spec="beast.evolution.branchratemodel.StrictClockModel"&gt;
            &lt;parameter dimension="1" estimate="false" id="clockRate" minordimension="1" name="clock.rate" value="1.0" /&gt;
        &lt;/branchRateModel&gt;

    &lt;/run&gt;
&lt;/beast&gt;
</pre>
</div>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Alexander E. Zarebski</p>
<p class="date">Created: 2025-01-29 Wed 13:42</p>
<p class="validation"><a href="https://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
