<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<!-- 2021-12-23 Thu 15:20 -->
<meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<title>aez-notes-beast2</title>
<meta name="author" content="Alex Zarebski" />
<meta name="generator" content="Org Mode" />
<style>
  #content { max-width: 60em; margin: auto; }
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #e6e6e6;
    border-radius: 3px;
    background-color: #f2f2f2;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: auto;
  }
  pre.src:before {
    display: none;
    position: absolute;
    top: -8px;
    right: 12px;
    padding: 3px;
    color: #555;
    background-color: #f2f2f299;
  }
  pre.src:hover:before { display: inline; margin-top: 14px;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-authinfo::before { content: 'Authinfo'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .equation-container {
    display: table;
    text-align: center;
    width: 100%;
  }
  .equation {
    vertical-align: middle;
  }
  .equation-label {
    display: table-cell;
    text-align: right;
    vertical-align: middle;
  }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
</style>
<link id="stylesheet" rel="stylesheet" type="text/css" href="../css/stylesheet-dark.css" />
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content" class="content">
<h1 class="title">aez-notes-beast2</h1>
<p>
<a href="../index.html">Home</a>
</p>

<button type="button" onclick="toggleStyle()">Light/Dark</button>
<script src="./main.js"></script>

<div id="table-of-contents" role="doc-toc">
<h2>Table of Contents</h2>
<div id="text-table-of-contents" role="doc-toc">
<ul>
<li><a href="#org3f61c1b">Links</a></li>
<li><a href="#org40ad059">Tools</a>
<ul>
<li><a href="#orga79c9df">Pretty printing Tracer output</a></li>
</ul>
</li>
<li><a href="#org791a8df">Examples</a>
<ul>
<li><a href="#org0cae390">BEAST2 as an MCMC engine: part I</a></li>
<li><a href="#org71098bf">BEAST2 as an MCMC engine: part II</a></li>
<li><a href="#org2764221">Birth-death simulation: Example I</a></li>
<li><a href="#org3af6932">Birth-death simulation: Example II</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-org3f61c1b" class="outline-2">
<h2 id="org3f61c1b">Links</h2>
<div class="outline-text-2" id="text-org3f61c1b">
<ul class="org-ul">
<li><a href="https://github.com/CompEvol/beast2">BEAST2 repository</a></li>
<li><a href="https://github.com/tgvaughan/feast">feast</a> a collection of resources from Tim Vaughan</li>
<li><a href="https://taming-the-beast.org/">Taming the BEAST</a> is a collection of tutorials for BEAST2</li>
<li><a href="https://github.com/laduplessis/skylinetools">skylinetools</a> is a package from Louis du Plessis which has skylining
functionality, (the <code>DateParser</code> constructor has a nifty trick.)</li>
</ul>
</div>
</div>

<div id="outline-container-org40ad059" class="outline-2">
<h2 id="org40ad059">Tools</h2>
<div class="outline-text-2" id="text-org40ad059">
<ul class="org-ul">
<li>You can run both <code>beast</code> and <code>beauti</code> from the command line as they are both
available in the <code>beast.jar</code>. In the case of <code>beauti</code> it will be something like
<code>java -sp &lt;path/to/beast.jar&gt; beast.app.beauti.Beauti</code>.</li>
<li>Tracer, FigTree and TreeAnnotator all comes as a separate jar files. You can
just click on the jar file and it should run the program correctly.</li>
</ul>
</div>

<div id="outline-container-orga79c9df" class="outline-3">
<h3 id="orga79c9df">Pretty printing Tracer output</h3>
<div class="outline-text-3" id="text-orga79c9df">
<p>
The following little script can be used to translate a summary produced by
Tracer into something that is easier to use in documents.
</p>

<div class="org-src-container">
<pre class="src src-R"><span style="color: #2aa1ae; background-color: #ecf3ec;">#</span><span style="color: #2aa1ae; background-color: #ecf3ec;">!/usr/bin/env Rscript</span>
<span style="color: #2aa1ae; background-color: #ecf3ec;"># </span><span style="color: #2aa1ae; background-color: #ecf3ec;">-*- mode:ess-mode; -*-</span>
<span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">#</span><span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">'</span>
<span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">#</span><span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">'</span><span style="color: #2aa1ae; background-color: #ecf3ec;"> tracer-table</span>
<span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">#</span><span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">'</span><span style="color: #2aa1ae; background-color: #ecf3ec;"> ============</span>
<span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">#</span><span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">'</span>
<span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">#</span><span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">'</span><span style="color: #2aa1ae; background-color: #ecf3ec;"> Tool to convert tracer summary to an org-mode/markdown style table.</span>
<span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">#</span><span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">'</span>
<span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">#</span><span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">'</span><span style="color: #2aa1ae; background-color: #ecf3ec;"> Usage</span>
<span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">#</span><span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">'</span><span style="color: #2aa1ae; background-color: #ecf3ec;"> -----</span>
<span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">#</span><span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">'</span>
<span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">#</span><span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">'</span><span style="color: #2aa1ae; background-color: #ecf3ec;"> ./tracer-table --input &lt;demo.txt&gt; --output demo.org</span>
<span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">#</span><span style="color: #2aa1ae; background-color: #ecf3ec; font-weight: bold;">'</span>

suppressPackageStartupMessages<span style="color: #3a81c3;">(</span><span style="color: #3a81c3;">library</span><span style="color: #6c3163;">(</span>argparse<span style="color: #6c3163;">)</span><span style="color: #3a81c3;">)</span>

parser <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> ArgumentParser<span style="color: #3a81c3;">()</span>

parser$add_argument<span style="color: #3a81c3;">(</span>
         <span style="color: #2d9574;">"-v"</span>,
         <span style="color: #2d9574;">"--verbose"</span>,
         action = <span style="color: #2d9574;">"store_true"</span>,
         default = <span style="color: #4e3163;">FALSE</span>,
         help = <span style="color: #2d9574;">"Verbose output"</span>
       <span style="color: #3a81c3;">)</span>
parser$add_argument<span style="color: #3a81c3;">(</span>
         <span style="color: #2d9574;">"-i"</span>,
         <span style="color: #2d9574;">"--input"</span>,
         type = <span style="color: #2d9574;">"character"</span>,
         help = <span style="color: #2d9574;">"Input file"</span>
       <span style="color: #3a81c3;">)</span>
parser$add_argument<span style="color: #3a81c3;">(</span>
         <span style="color: #2d9574;">"-o"</span>,
         <span style="color: #2d9574;">"--output"</span>,
         type = <span style="color: #2d9574;">"character"</span>,
         help = <span style="color: #2d9574;">"Output file"</span>
       <span style="color: #3a81c3;">)</span>

args <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> parser$parse_args<span style="color: #3a81c3;">()</span>


<span style="color: #6c3163; font-weight: bold;">main</span> <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> <span style="color: #3a81c3;">function</span><span style="color: #3a81c3;">(</span>args<span style="color: #3a81c3;">)</span> <span style="color: #3a81c3;">{</span>
  foo <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> readLines<span style="color: #6c3163;">(</span>args$input<span style="color: #6c3163;">)</span> |&gt;
    gsub<span style="color: #6c3163;">(</span>pattern = <span style="color: #2d9574;">"\t"</span>, replacement = <span style="color: #2d9574;">" | "</span><span style="color: #6c3163;">)</span> |&gt;
    gsub<span style="color: #6c3163;">(</span>pattern = <span style="color: #2d9574;">"^"</span>, replacement = <span style="color: #2d9574;">"| "</span><span style="color: #6c3163;">)</span> |&gt;
    gsub<span style="color: #6c3163;">(</span>pattern = <span style="color: #2d9574;">"$"</span>, replacement = <span style="color: #2d9574;">" |"</span><span style="color: #6c3163;">)</span>
  tbl_header <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> head<span style="color: #6c3163;">(</span>foo, <span style="color: #4e3163;">1</span><span style="color: #6c3163;">)</span>
  tbl_hline <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> tbl_header |&gt;
    gsub<span style="color: #6c3163;">(</span>pattern = <span style="color: #2d9574;">"[_\ a-zA-Z]"</span>, replacement = <span style="color: #2d9574;">"-"</span><span style="color: #6c3163;">)</span> |&gt;
    gsub<span style="color: #6c3163;">(</span>pattern = <span style="color: #2d9574;">"-\\|-"</span>, replacement = <span style="color: #2d9574;">"-+-"</span><span style="color: #6c3163;">)</span>
  tbl_body <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> tail<span style="color: #6c3163;">(</span>foo, -<span style="color: #4e3163;">1</span><span style="color: #6c3163;">)</span>
  bar <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> c<span style="color: #6c3163;">(</span>tbl_header, tbl_hline, tbl_body<span style="color: #6c3163;">)</span>
  writeLines<span style="color: #6c3163;">(</span>text = bar, con = args$output<span style="color: #6c3163;">)</span>
<span style="color: #3a81c3;">}</span>

<span style="color: #3a81c3;">if</span> <span style="color: #3a81c3;">(</span>!interactive<span style="color: #6c3163;">()</span><span style="color: #3a81c3;">)</span> <span style="color: #3a81c3;">{</span>
  args <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> parser$parse_args<span style="color: #6c3163;">()</span>
  main<span style="color: #6c3163;">(</span>args<span style="color: #6c3163;">)</span>
<span style="color: #3a81c3;">}</span>
</pre>
</div>
</div>
</div>
</div>


<div id="outline-container-org791a8df" class="outline-2">
<h2 id="org791a8df">Examples</h2>
<div class="outline-text-2" id="text-org791a8df">
<ul class="org-ul">
<li><a href="#org0cae390">BEAST2 as an MCMC engine: part I</a></li>
<li><a href="#org71098bf">BEAST2 as an MCMC engine: part II</a></li>
<li><a href="#org2764221">Birth-death simulation: Example I</a></li>
<li><a href="#org3af6932">Birth-death simulation: Example II</a></li>
</ul>
</div>

<div id="outline-container-org0cae390" class="outline-3">
<h3 id="org0cae390">BEAST2 as an MCMC engine: part I</h3>
<div class="outline-text-3" id="text-org0cae390">
<p>
This example looks at how to use BEAST2 to simulate from a standard normal
distribution using MCMC. The XML for this has the following structure
</p>

<div class="org-src-container">
<pre class="src src-xml">&lt;?<span style="color: #3a81c3;">xml</span> <span style="color: #da8b55;">version="1.0" encoding="UTF-8" standalone="no"</span>?&gt;
&lt;<span style="color: #6c3163; font-weight: bold;">beast</span> <span style="color: #715ab1;">version</span>=<span style="color: #2d9574;">"2.0"</span>&gt;

  &lt;&lt;<span style="color: #6c3163; font-weight: bold;">targetDistribution</span>&gt;&gt;

  &lt;&lt;<span style="color: #6c3163; font-weight: bold;">mcmc</span>&gt;&gt;

&lt;/<span style="color: #6c3163; font-weight: bold;">beast</span>&gt;
</pre>
</div>

<p>
The <a href="#org94fd60f"><code>targetDistribution</code></a> specifies the distribution we will sample from and the <a href="#orgb641241">mcmc</a> specifies how details of the MCMC algorithm.
</p>
</div>

<div id="outline-container-org94fd60f" class="outline-4">
<h4 id="org94fd60f">The target distribution</h4>
<div class="outline-text-4" id="text-org94fd60f">
<p>
We use a <code>Prior</code> to represent the distribution because this makes it easy to
evaluate the distribution at a particular value.
</p>

<div class="org-src-container">
<pre class="src src-xml" id="org4371404">&lt;<span style="color: #6c3163; font-weight: bold;">distribution</span> <span style="color: #715ab1;">id</span>=<span style="color: #2d9574;">"posterior"</span> <span style="color: #715ab1;">spec</span>=<span style="color: #2d9574;">"beast.math.distributions.Prior"</span> <span style="color: #715ab1;">x</span>=<span style="color: #2d9574;">"@x"</span>&gt;
  &lt;<span style="color: #6c3163; font-weight: bold;">distr</span> <span style="color: #715ab1;">spec</span>=<span style="color: #2d9574;">"beast.math.distributions.Normal"</span> <span style="color: #715ab1;">mean</span>=<span style="color: #2d9574;">"0.0"</span> <span style="color: #715ab1;">sigma</span>=<span style="color: #2d9574;">"1.0"</span> /&gt;
&lt;/<span style="color: #6c3163; font-weight: bold;">distribution</span>&gt;
</pre>
</div>
</div>
</div>

<div id="outline-container-orgb641241" class="outline-4">
<h4 id="orgb641241">MCMC</h4>
<div class="outline-text-4" id="text-orgb641241">
<p>
We use a <code>run</code> tag to specify the MCMC. This requires a <code>state</code> object to
describe the parameters we are sampling, the <code>distribution</code> over the state, and
an <code>operator</code> and <code>logger</code> to perturb the state and record the results.
</p>

<div class="org-src-container">
<pre class="src src-xml" id="orgb21e102">&lt;<span style="color: #6c3163; font-weight: bold;">run</span> <span style="color: #715ab1;">chainLength</span>=<span style="color: #2d9574;">"10000000"</span> <span style="color: #715ab1;">id</span>=<span style="color: #2d9574;">"mcmc"</span> <span style="color: #715ab1;">preBurnin</span>=<span style="color: #2d9574;">"100"</span> <span style="color: #715ab1;">spec</span>=<span style="color: #2d9574;">"beast.core.MCMC"</span>&gt;

    &lt;<span style="color: #6c3163; font-weight: bold;">state</span> <span style="color: #715ab1;">id</span>=<span style="color: #2d9574;">"state"</span> <span style="color: #715ab1;">storeEvery</span>=<span style="color: #2d9574;">"1000"</span>&gt;
        &lt;<span style="color: #6c3163; font-weight: bold;">parameter</span> <span style="color: #715ab1;">estimate</span>=<span style="color: #2d9574;">"true"</span> <span style="color: #715ab1;">id</span>=<span style="color: #2d9574;">"x"</span> <span style="color: #715ab1;">name</span>=<span style="color: #2d9574;">"stateNode"</span> <span style="color: #715ab1;">value</span>=<span style="color: #2d9574;">"1.0"</span> /&gt;
    &lt;/<span style="color: #6c3163; font-weight: bold;">state</span>&gt;

    &lt;<span style="color: #6c3163; font-weight: bold;">distribution</span> <span style="color: #715ab1;">idref</span>=<span style="color: #2d9574;">"posterior"</span> /&gt;

    &lt;<span style="color: #6c3163; font-weight: bold;">operator</span> <span style="color: #715ab1;">spec</span>=<span style="color: #2d9574;">"beast.evolution.operators.RealRandomWalkOperator"</span>
              <span style="color: #715ab1;">windowSize</span>=<span style="color: #2d9574;">"0.1"</span>
              <span style="color: #715ab1;">useGaussian</span>=<span style="color: #2d9574;">"true"</span>
              <span style="color: #715ab1;">weight</span>=<span style="color: #2d9574;">"1.0"</span>
              <span style="color: #715ab1;">parameter</span>=<span style="color: #2d9574;">"@x"</span> /&gt;

    &lt;<span style="color: #6c3163; font-weight: bold;">logger</span> <span style="color: #715ab1;">id</span>=<span style="color: #2d9574;">"screenlog"</span> <span style="color: #715ab1;">logEvery</span>=<span style="color: #2d9574;">"1000000"</span>&gt;
        &lt;<span style="color: #6c3163; font-weight: bold;">log</span> <span style="color: #715ab1;">idref</span>=<span style="color: #2d9574;">"x"</span> /&gt;
    &lt;/<span style="color: #6c3163; font-weight: bold;">logger</span>&gt;

    &lt;<span style="color: #6c3163; font-weight: bold;">logger</span> <span style="color: #715ab1;">fileName</span>=<span style="color: #2d9574;">"beast.csv"</span>
            <span style="color: #715ab1;">id</span>=<span style="color: #2d9574;">"tracelog"</span>
            <span style="color: #715ab1;">logEvery</span>=<span style="color: #2d9574;">"10000"</span>
            <span style="color: #715ab1;">model</span>=<span style="color: #2d9574;">"@posterior"</span>&gt;
        &lt;<span style="color: #6c3163; font-weight: bold;">log</span> <span style="color: #715ab1;">idref</span>=<span style="color: #2d9574;">"posterior"</span> /&gt;
        &lt;<span style="color: #6c3163; font-weight: bold;">log</span> <span style="color: #715ab1;">idref</span>=<span style="color: #2d9574;">"x"</span> /&gt;
    &lt;/<span style="color: #6c3163; font-weight: bold;">logger</span>&gt;

&lt;/<span style="color: #6c3163; font-weight: bold;">run</span>&gt;
</pre>
</div>
</div>
</div>

<div id="outline-container-orgea2b6e0" class="outline-4">
<h4 id="orgea2b6e0">Running the sampler</h4>
<div class="outline-text-4" id="text-orgea2b6e0">
<p>
Here is the command to actually run this program.
</p>

<div class="org-src-container">
<pre class="src src-sh">java -jar beast.jar -seed <span style="color: #4e3163;">1</span> -overwrite example-01.xml
</pre>
</div>

<p>
I have used the <code>-overwrite</code> flag because otherwise it will prompt you each time
unless you change the output file name.
</p>
</div>
</div>

<div id="outline-container-org926e6bd" class="outline-4">
<h4 id="org926e6bd">Visualising the results</h4>
<div class="outline-text-4" id="text-org926e6bd">
<p>
Here is a visualisation of the resulting samples along with the distribution we
were sampling from.
</p>


<div id="org47c51c4" class="figure">
<p><img src="../images/beast-demo.png" alt="beast-demo.png" width="800px" />
</p>
</div>

<div class="org-src-container">
<pre class="src src-R"><span style="color: #3a81c3;">library</span><span style="color: #3a81c3;">(</span>ggplot2<span style="color: #3a81c3;">)</span>
<span style="color: #3a81c3;">library</span><span style="color: #3a81c3;">(</span>cowplot<span style="color: #3a81c3;">)</span>

beast_samples <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> read.table<span style="color: #3a81c3;">(</span><span style="color: #2d9574;">"beast.csv"</span>,
  sep = <span style="color: #2d9574;">"\t"</span>,
  comment.char = <span style="color: #2d9574;">"#"</span>,
  header = <span style="color: #4e3163;">TRUE</span>
<span style="color: #3a81c3;">)</span>

g1 <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> ggplot<span style="color: #3a81c3;">()</span> +
  geom_line<span style="color: #3a81c3;">(</span>
    data = beast_samples,
    mapping = aes<span style="color: #6c3163;">(</span>x = Sample, y = x<span style="color: #6c3163;">)</span>
  <span style="color: #3a81c3;">)</span> +
  labs<span style="color: #3a81c3;">(</span>x = <span style="color: #2d9574;">"Sample"</span>, y = <span style="color: #2d9574;">"X"</span><span style="color: #3a81c3;">)</span> +
  theme_classic<span style="color: #3a81c3;">()</span>

g2 <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> ggplot<span style="color: #3a81c3;">()</span> +
  geom_histogram<span style="color: #3a81c3;">(</span>
    data = beast_samples,
    mapping = aes<span style="color: #6c3163;">(</span>x = x, y = ..density..<span style="color: #6c3163;">)</span>,
    bins = <span style="color: #4e3163;">10</span>
  <span style="color: #3a81c3;">)</span> +
  stat_function<span style="color: #3a81c3;">(</span>
    fun = \<span style="color: #6c3163;">(</span>x<span style="color: #6c3163;">)</span> dnorm<span style="color: #6c3163;">(</span>x<span style="color: #6c3163;">)</span>,
    geom = <span style="color: #2d9574;">"line"</span>
  <span style="color: #3a81c3;">)</span> +
  labs<span style="color: #3a81c3;">(</span>x = <span style="color: #2d9574;">"X"</span>, y = <span style="color: #2d9574;">"Density"</span><span style="color: #3a81c3;">)</span> +
  theme_classic<span style="color: #3a81c3;">()</span>

g <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> plot_grid<span style="color: #3a81c3;">(</span>g1, g2<span style="color: #3a81c3;">)</span>

ggsave<span style="color: #3a81c3;">(</span><span style="color: #2d9574;">"beast-demo.png"</span>,
       g,
       height = <span style="color: #4e3163;">10.5</span>,
       width = <span style="color: #4e3163;">14.8</span>,
       units = <span style="color: #2d9574;">"cm"</span><span style="color: #3a81c3;">)</span>
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-org71098bf" class="outline-3">
<h3 id="org71098bf">BEAST2 as an MCMC engine: part II</h3>
<div class="outline-text-3" id="text-org71098bf">
<p>
This example looks at estimating the probability of heads in coin flips. We
start with at beta prior distribution and take values of the number of observed
heads and tails from the command line.
</p>

<div class="org-src-container">
<pre class="src src-xml">&lt;?<span style="color: #3a81c3;">xml</span> <span style="color: #da8b55;">version="1.0" encoding="UTF-8" standalone="no"</span>?&gt;
&lt;<span style="color: #6c3163; font-weight: bold;">beast</span> <span style="color: #715ab1;">version</span>=<span style="color: #2d9574;">"2.0"</span>&gt;

    &lt;<span style="color: #6c3163; font-weight: bold;">distribution</span> <span style="color: #715ab1;">id</span>=<span style="color: #2d9574;">"posterior"</span> <span style="color: #715ab1;">spec</span>=<span style="color: #2d9574;">"beast.core.util.CompoundDistribution"</span> &gt;
        &lt;<span style="color: #6c3163; font-weight: bold;">distribution</span> <span style="color: #715ab1;">id</span>=<span style="color: #2d9574;">"prior"</span> <span style="color: #715ab1;">spec</span>=<span style="color: #2d9574;">"beast.math.distributions.Prior"</span> <span style="color: #715ab1;">x</span>=<span style="color: #2d9574;">"@p"</span>&gt;
            &lt;<span style="color: #6c3163; font-weight: bold;">distr</span> <span style="color: #715ab1;">spec</span>=<span style="color: #2d9574;">"beast.math.distributions.Beta"</span> <span style="color: #715ab1;">alpha</span>=<span style="color: #2d9574;">"0.5"</span> <span style="color: #715ab1;">beta</span>=<span style="color: #2d9574;">"0.5"</span> /&gt;
        &lt;/<span style="color: #6c3163; font-weight: bold;">distribution</span>&gt;

        &lt;<span style="color: #6c3163; font-weight: bold;">distribution</span> <span style="color: #715ab1;">id</span>=<span style="color: #2d9574;">"likelihood"</span> <span style="color: #715ab1;">spec</span>=<span style="color: #2d9574;">"beast.math.distributions.Prior"</span> <span style="color: #715ab1;">x</span>=<span style="color: #2d9574;">"@p"</span>&gt;
            &lt;<span style="color: #6c3163; font-weight: bold;">distr</span> <span style="color: #715ab1;">spec</span>=<span style="color: #2d9574;">"beast.math.distributions.Beta"</span> <span style="color: #715ab1;">alpha</span>=<span style="color: #2d9574;">"$(heads)"</span> <span style="color: #715ab1;">beta</span>=<span style="color: #2d9574;">"$(tails)"</span> /&gt;
        &lt;/<span style="color: #6c3163; font-weight: bold;">distribution</span>&gt;
    &lt;/<span style="color: #6c3163; font-weight: bold;">distribution</span>&gt;

    &lt;<span style="color: #6c3163; font-weight: bold;">run</span> <span style="color: #715ab1;">chainLength</span>=<span style="color: #2d9574;">"10000000"</span> <span style="color: #715ab1;">id</span>=<span style="color: #2d9574;">"mcmc"</span> <span style="color: #715ab1;">preBurnin</span>=<span style="color: #2d9574;">"100"</span> <span style="color: #715ab1;">spec</span>=<span style="color: #2d9574;">"beast.core.MCMC"</span>&gt;

        &lt;<span style="color: #6c3163; font-weight: bold;">state</span> <span style="color: #715ab1;">id</span>=<span style="color: #2d9574;">"state"</span> <span style="color: #715ab1;">storeEvery</span>=<span style="color: #2d9574;">"1000"</span>&gt;
            &lt;<span style="color: #6c3163; font-weight: bold;">parameter</span> <span style="color: #715ab1;">estimate</span>=<span style="color: #2d9574;">"true"</span> <span style="color: #715ab1;">id</span>=<span style="color: #2d9574;">"p"</span> <span style="color: #715ab1;">name</span>=<span style="color: #2d9574;">"stateNode"</span> <span style="color: #715ab1;">value</span>=<span style="color: #2d9574;">"0.5"</span> /&gt;
        &lt;/<span style="color: #6c3163; font-weight: bold;">state</span>&gt;

        &lt;<span style="color: #6c3163; font-weight: bold;">distribution</span> <span style="color: #715ab1;">idref</span>=<span style="color: #2d9574;">"posterior"</span> /&gt;

        &lt;<span style="color: #6c3163; font-weight: bold;">operator</span> <span style="color: #715ab1;">spec</span>=<span style="color: #2d9574;">"beast.evolution.operators.RealRandomWalkOperator"</span> <span style="color: #715ab1;">windowSize</span>=<span style="color: #2d9574;">"0.1"</span> <span style="color: #715ab1;">useGaussian</span>=<span style="color: #2d9574;">"true"</span> <span style="color: #715ab1;">weight</span>=<span style="color: #2d9574;">"1.0"</span> <span style="color: #715ab1;">parameter</span>=<span style="color: #2d9574;">"@p"</span> /&gt;

        &lt;<span style="color: #6c3163; font-weight: bold;">logger</span> <span style="color: #715ab1;">id</span>=<span style="color: #2d9574;">"screenlog"</span> <span style="color: #715ab1;">logEvery</span>=<span style="color: #2d9574;">"1000000"</span>&gt;
            &lt;<span style="color: #6c3163; font-weight: bold;">log</span> <span style="color: #715ab1;">idref</span>=<span style="color: #2d9574;">"p"</span> /&gt;
        &lt;/<span style="color: #6c3163; font-weight: bold;">logger</span>&gt;

        &lt;<span style="color: #6c3163; font-weight: bold;">logger</span> <span style="color: #715ab1;">fileName</span>=<span style="color: #2d9574;">"beast-02.csv"</span> <span style="color: #715ab1;">id</span>=<span style="color: #2d9574;">"tracelog"</span> <span style="color: #715ab1;">logEvery</span>=<span style="color: #2d9574;">"10000"</span> <span style="color: #715ab1;">model</span>=<span style="color: #2d9574;">"@posterior"</span>&gt;
            &lt;<span style="color: #6c3163; font-weight: bold;">log</span> <span style="color: #715ab1;">idref</span>=<span style="color: #2d9574;">"posterior"</span> /&gt;
            &lt;<span style="color: #6c3163; font-weight: bold;">log</span> <span style="color: #715ab1;">idref</span>=<span style="color: #2d9574;">"p"</span> /&gt;
        &lt;/<span style="color: #6c3163; font-weight: bold;">logger</span>&gt;

    &lt;/<span style="color: #6c3163; font-weight: bold;">run</span>&gt;

&lt;/<span style="color: #6c3163; font-weight: bold;">beast</span>&gt;
</pre>
</div>
</div>

<div id="outline-container-orgd6ccc14" class="outline-4">
<h4 id="orgd6ccc14">Running the sampler</h4>
<div class="outline-text-4" id="text-orgd6ccc14">
<p>
Here is the command to actually run this program.
</p>

<div class="org-src-container">
<pre class="src src-sh">java -jar beast.jar -seed <span style="color: #4e3163;">1</span> -overwrite -D <span style="color: #2d9574;">"heads=3,tails=7"</span> example-02.xml
</pre>
</div>

<p>
The <code>-D</code> flag has been used to specify how many times the coin came up heads and
tails.
</p>
</div>
</div>

<div id="outline-container-org17a4df2" class="outline-4">
<h4 id="org17a4df2">Visualising the results</h4>
<div class="outline-text-4" id="text-org17a4df2">
<p>
Here is a visualisation of the resulting samples along with the distribution we
were sampling from.
</p>


<div id="org2847a88" class="figure">
<p><img src="../images/beast-02.png" alt="beast-02.png" width="800px" />
</p>
</div>

<div class="org-src-container">
<pre class="src src-R"><span style="color: #3a81c3;">library</span><span style="color: #3a81c3;">(</span>ggplot2<span style="color: #3a81c3;">)</span>
<span style="color: #3a81c3;">library</span><span style="color: #3a81c3;">(</span>cowplot<span style="color: #3a81c3;">)</span>

beast_samples <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> read.table<span style="color: #3a81c3;">(</span><span style="color: #2d9574;">"beast-02.csv"</span>,
  sep = <span style="color: #2d9574;">"\t"</span>,
  comment.char = <span style="color: #2d9574;">"#"</span>,
  header = <span style="color: #4e3163;">TRUE</span>
<span style="color: #3a81c3;">)</span>

g1 <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> ggplot<span style="color: #3a81c3;">()</span> +
  geom_line<span style="color: #3a81c3;">(</span>
    data = beast_samples,
    mapping = aes<span style="color: #6c3163;">(</span>x = Sample, y = p<span style="color: #6c3163;">)</span>
  <span style="color: #3a81c3;">)</span> +
  labs<span style="color: #3a81c3;">(</span>x = <span style="color: #2d9574;">"Sample"</span>, y = <span style="color: #2d9574;">"p"</span><span style="color: #3a81c3;">)</span> +
  theme_classic<span style="color: #3a81c3;">()</span>

g2 <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> ggplot<span style="color: #3a81c3;">()</span> +
  geom_histogram<span style="color: #3a81c3;">(</span>
    data = beast_samples,
    mapping = aes<span style="color: #6c3163;">(</span>x = p, y = ..density..<span style="color: #6c3163;">)</span>,
    bins = <span style="color: #4e3163;">15</span>
  <span style="color: #3a81c3;">)</span> +
  stat_function<span style="color: #3a81c3;">(</span>
    fun = \<span style="color: #6c3163;">(</span>x<span style="color: #6c3163;">)</span> dbeta<span style="color: #6c3163;">(</span>x, <span style="color: #4e3163;">0.5</span> + <span style="color: #4e3163;">3</span>, <span style="color: #4e3163;">0.5</span> + <span style="color: #4e3163;">7</span><span style="color: #6c3163;">)</span>,
    geom = <span style="color: #2d9574;">"line"</span>
  <span style="color: #3a81c3;">)</span> +
  labs<span style="color: #3a81c3;">(</span>x = <span style="color: #2d9574;">"X"</span>, y = <span style="color: #2d9574;">"Density"</span><span style="color: #3a81c3;">)</span> +
  theme_classic<span style="color: #3a81c3;">()</span>

g <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> plot_grid<span style="color: #3a81c3;">(</span>g1, g2<span style="color: #3a81c3;">)</span>

ggsave<span style="color: #3a81c3;">(</span><span style="color: #2d9574;">"beast-02.png"</span>,
       g,
       height = <span style="color: #4e3163;">10.5</span>,
       width = <span style="color: #4e3163;">14.8</span>,
       units = <span style="color: #2d9574;">"cm"</span><span style="color: #3a81c3;">)</span>
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-org2764221" class="outline-3">
<h3 id="org2764221">Birth-death simulation: Example I</h3>
<div class="outline-text-3" id="text-org2764221">
<p>
In this example we will simulate sequences on a realisation of the birth-death
process and then attempt to estimate the birth rate from these sequences. We
will use <code>ape</code> to simulate the birth-death process and <code>phangorn</code> to simulate the
sequences. To ensure a correct XML specification we will use <code>beauti</code> to generate
the XML for the analysis. Finally, we will use <code>tracer</code> to check the results.
</p>
</div>

<div id="outline-container-orgb54dba1" class="outline-4">
<h4 id="orgb54dba1">Simulating the tree</h4>
<div class="outline-text-4" id="text-orgb54dba1">
<p>
We will use years as our unit of time. Assuming that an individual is infectious
for 7 days (\(1/52\) of a year) on average, and that \(1\%\) of infections are
sequenced we end up with the following equations for the rates: \(\mu + \psi =
52\) and \(\psi / (\psi + \mu) = 1 / 100\). Solving these give us \(\mu =
5148/100\) and \(\psi = 52/100\). If we have an \(\lambda / (\mu + \psi) = R_{0}
= 2\) then \(\lambda = 104\).
</p>

<div class="org-src-container">
<pre class="src src-R">birth_rate <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> <span style="color: #4e3163;">104</span>
death_rate <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> <span style="color: #4e3163;">5148</span> / <span style="color: #4e3163;">100</span>
sampling_rate <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> <span style="color: #4e3163;">52</span> / <span style="color: #4e3163;">100</span>
sim_duration <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> <span style="color: #4e3163;">0.15</span> <span style="color: #2aa1ae; background-color: #ecf3ec;"># </span><span style="color: #2aa1ae; background-color: #ecf3ec;">year</span>
</pre>
</div>

<p>
The <code>rlineage</code> function returns a realisation of the birth-death process as a
<code>phylo</code> object.
</p>

<div class="org-src-container">
<pre class="src src-R" id="orgc0fc2de">net_removal_rate <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> death_rate + sampling_rate
sim_tree <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> rlineage<span style="color: #3a81c3;">(</span>birth_rate,
                     net_removal_rate,
                     Tmax = sim_duration<span style="color: #3a81c3;">)</span>
</pre>
</div>

<p>
We include the tip dates as part of the tip labels because it makes it much
easier to read the data into Beauti if you can parse this information from the
top labels.
</p>

<div class="org-src-container">
<pre class="src src-R" id="org3949ca8">num_tips <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> length<span style="color: #3a81c3;">(</span>sim_tree$tip.label<span style="color: #3a81c3;">)</span>
tip_fwd_times <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> node.depth.edgelength<span style="color: #3a81c3;">(</span>sim_tree<span style="color: #3a81c3;">)[</span><span style="color: #4e3163;">1</span>:num_tips<span style="color: #3a81c3;">]</span>
sim_tree$tip.label <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> sprintf<span style="color: #3a81c3;">(</span><span style="color: #2d9574;">"%s_%f"</span>,
                              sim_tree$tip.label,
                              <span style="color: #4e3163;">2021</span> + tip_fwd_times<span style="color: #3a81c3;">)</span>
</pre>
</div>

<p>
The reconstructed tree is the tree that remains when we sample the extinct tips
(ie those that are not extant). The probability that a tip is sequenced is
\(\psi / (\mu + \psi\)\).
</p>

<div class="org-src-container">
<pre class="src src-R" id="orge1f770e">extant_mask <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> tip_fwd_times &gt;= sim_duration
num_removed <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> sum<span style="color: #3a81c3;">(</span>!extant_mask<span style="color: #3a81c3;">)</span>
num_sampled <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> rbinom<span style="color: #3a81c3;">(</span>n = <span style="color: #4e3163;">1</span>,
                      size = num_removed,
                      prob = sampling_rate / net_removal_rate<span style="color: #3a81c3;">)</span>
sampled_tips <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> sample<span style="color: #3a81c3;">(</span>x = sim_tree$tip.label<span style="color: #6c3163;">[</span>!extant_mask<span style="color: #6c3163;">]</span>,
                       size = num_sampled<span style="color: #3a81c3;">)</span>
reconstructed_tree <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> keep.tip<span style="color: #3a81c3;">(</span>phy = sim_tree,
                               tip = sampled_tips<span style="color: #3a81c3;">)</span>
</pre>
</div>

<p>
We can combine all of this into a function
</p>

<div class="org-src-container">
<pre class="src src-R"><span style="color: #6c3163; font-weight: bold;">rand_reconstructed_tree</span> <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> <span style="color: #3a81c3;">function</span><span style="color: #3a81c3;">(</span>birth_rate,
                                    death_rate,
                                    sampling_rate,
                                    duration<span style="color: #3a81c3;">)</span> <span style="color: #3a81c3;">{</span>
  net_removal_rate <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> death_rate + sampling_rate
  sim_tree <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> rlineage<span style="color: #6c3163;">(</span>birth_rate,
                       net_removal_rate,
                       Tmax = sim_duration<span style="color: #6c3163;">)</span>

  num_tips <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> length<span style="color: #6c3163;">(</span>sim_tree$tip.label<span style="color: #6c3163;">)</span>
  tip_fwd_times <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> node.depth.edgelength<span style="color: #6c3163;">(</span>sim_tree<span style="color: #6c3163;">)[</span><span style="color: #4e3163;">1</span>:num_tips<span style="color: #6c3163;">]</span>
  sim_tree$tip.label <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> sprintf<span style="color: #6c3163;">(</span><span style="color: #2d9574;">"%s_%f"</span>,
                                sim_tree$tip.label,
                                <span style="color: #4e3163;">2021</span> + tip_fwd_times<span style="color: #6c3163;">)</span>

  extant_mask <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> tip_fwd_times &gt;= sim_duration
  num_removed <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> sum<span style="color: #6c3163;">(</span>!extant_mask<span style="color: #6c3163;">)</span>
  num_sampled <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> rbinom<span style="color: #6c3163;">(</span>n = <span style="color: #4e3163;">1</span>,
                        size = num_removed,
                        prob = sampling_rate / net_removal_rate<span style="color: #6c3163;">)</span>
  sampled_tips <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> sample<span style="color: #6c3163;">(</span>x = sim_tree$tip.label<span style="color: #2d9574;">[</span>!extant_mask<span style="color: #2d9574;">]</span>,
                         size = num_sampled<span style="color: #6c3163;">)</span>
  reconstructed_tree <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> keep.tip<span style="color: #6c3163;">(</span>phy = sim_tree,
                                 tip = sampled_tips<span style="color: #6c3163;">)</span>

  <span style="color: #3a81c3;">return</span><span style="color: #6c3163;">(</span>reconstructed_tree<span style="color: #6c3163;">)</span>
<span style="color: #3a81c3;">}</span>
</pre>
</div>

<p>
We want a simulation that has a reasonable number of sequences in it, so we
generate random trees until we have one with at least \(20\) tips.
</p>

<div class="org-src-container">
<pre class="src src-R">iter_count <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> <span style="color: #4e3163;">0</span>
num_seqs <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> <span style="color: #4e3163;">0</span>
<span style="color: #3a81c3;">while</span> <span style="color: #3a81c3;">(</span>iter_count &lt; <span style="color: #4e3163;">100</span> &amp;&amp; num_seqs &lt; <span style="color: #4e3163;">20</span><span style="color: #3a81c3;">)</span> <span style="color: #3a81c3;">{</span>
  reconstructed_tree <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> rand_reconstructed_tree<span style="color: #6c3163;">(</span>birth_rate,
                                                death_rate,
                                                sampling_rate,
                                                sim_duration<span style="color: #6c3163;">)</span>
  iter_count <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> iter_count + <span style="color: #4e3163;">1</span>
  num_seqs <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> length<span style="color: #6c3163;">(</span>reconstructed_tree$tip.label<span style="color: #6c3163;">)</span>
  print<span style="color: #6c3163;">(</span>iter_count<span style="color: #6c3163;">)</span>
<span style="color: #3a81c3;">}</span>
</pre>
</div>

<p>
This tree should be saved for later use.
</p>

<div class="org-src-container">
<pre class="src src-R">write.tree<span style="color: #3a81c3;">(</span>phy = reconstructed_tree,
           file = <span style="color: #2d9574;">"reconstructed-tree.newick"</span><span style="color: #3a81c3;">)</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-orga788b22" class="outline-4">
<h4 id="orga788b22">Simulating the sequences</h4>
<div class="outline-text-4" id="text-orga788b22">
<p>
Then we use the <code>simSeq</code> function to simulate a set of genomes for the leaves of
this tree. By default this function uses the JC model for the sequence
simulation with a rate of 1 (using the branch lengths as the units of time). RNA
viruses have a mutation rate of \(\approx 10^{-3}\) substitutions per site per
year. We are already working in units of years so we can use this as our rate.
The length of the sequence we will use is 2000 which is similar to the HA of
influenza.
</p>

<div class="org-src-container">
<pre class="src src-R">sim_seqs <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> simSeq<span style="color: #3a81c3;">(</span>x = reconstructed_tree,
                   rate = <span style="color: #4e3163;">1e</span>-<span style="color: #4e3163;">3</span>,
                   l = <span style="color: #4e3163;">2000</span>,
                   type = <span style="color: #2d9574;">"DNA"</span><span style="color: #3a81c3;">)</span>
</pre>
</div>

<p>
Then we write this to a file so we can read it into Beauti.
</p>

<div class="org-src-container">
<pre class="src src-R">write.phyDat<span style="color: #3a81c3;">(</span>x = sim_seqs,
             file = <span style="color: #2d9574;">"sequences.fasta"</span>,
             format = <span style="color: #2d9574;">"fasta"</span><span style="color: #3a81c3;">)</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-org175bd58" class="outline-4">
<h4 id="org175bd58">Constructing the XML</h4>
<div class="outline-text-4" id="text-org175bd58">
<div class="org-src-container">
<pre class="src src-sh">java -cp ~/Documents/beast2/build/dist/beast.jar beast.app.beauti.Beauti
</pre>
</div>


<ol class="org-ol">
<li>Open Beauti and load in the alignment produced by the simulation above.</li>
<li>Open the <b>Tip Dates</b> tab and auto-configure to get the tip dates.
<ul class="org-ul">
<li>Since the sequences have been simulated with JC model with rate 1 nothing
special is needed for the <b>Site Model</b> and <b>Clock Model</b> tabs.</li>
</ul></li>
<li>Open the <b>Priors</b> tab and select the <b>Birth Death Skyline Serial</b> prior
<ul class="org-ul">
<li>The reproduction number has dimension 10 by default, we used constant rates
so this should be reduced to 1.</li>
<li>Go over the parameters and make sure that for each one the prior looks
sensible. To make things a bit easier, set informative priors for each
parameter.</li>
</ul></li>
<li>Save the analysis as an XML.</li>
</ol>
</div>
</div>

<div id="outline-container-orgfb60d58" class="outline-4">
<h4 id="orgfb60d58">Running the sampler</h4>
<div class="outline-text-4" id="text-orgfb60d58">
<p>
Running the sampler for \(10^{6}\) iterations does not take very long and allows
us to see which variables are getting stuck. There are some suggested
adjustments to the operators which should be done before increasing the sample
size to \(10^{7}\).
</p>
</div>
</div>

<div id="outline-container-orgd2ff327" class="outline-4">
<h4 id="orgd2ff327">Analysing the posterior samples</h4>
<div class="outline-text-4" id="text-orgd2ff327">
<p>
Tracer can be used for preliminary investigation of the posterior samples as
shown in the following table (generated using <a href="#orga79c9df"><code>tracer-table</code></a>).
</p>

<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">


<colgroup>
<col  class="org-left" />

<col  class="org-right" />

<col  class="org-left" />

<col  class="org-right" />
</colgroup>
<thead>
<tr>
<th scope="col" class="org-left">Summary Statistic</th>
<th scope="col" class="org-right">Mean</th>
<th scope="col" class="org-left">95% HPD Interval</th>
<th scope="col" class="org-right">Effective Sample Size</th>
</tr>
</thead>
<tbody>
<tr>
<td class="org-left">clock rate</td>
<td class="org-right">7.641E-4</td>
<td class="org-left">[3.9092E-4, 1.171E-3]</td>
<td class="org-right">496.5</td>
</tr>

<tr>
<td class="org-left">origin time</td>
<td class="org-right">0.1509</td>
<td class="org-left">[0.1146, 0.1935]</td>
<td class="org-right">254.7</td>
</tr>

<tr>
<td class="org-left">become uninfectious rate</td>
<td class="org-right">51.9713</td>
<td class="org-left">[50.0723, 53.9649]</td>
<td class="org-right">4406.5</td>
</tr>

<tr>
<td class="org-left">reproductive number</td>
<td class="org-right">2.0901</td>
<td class="org-left">[1.8752, 2.3189]</td>
<td class="org-right">581.7</td>
</tr>

<tr>
<td class="org-left">sampling proportion</td>
<td class="org-right">0.0204</td>
<td class="org-left">[1.9394E-3, 0.0435]</td>
<td class="org-right">116.5</td>
</tr>
</tbody>
</table>

<p>
To display the prior and posteriors on the same figure we need to resort to
<code>ggplot2</code> though. There appear to be some identifiability issues (which is to be
expected), but the estimates of the origin and basic reproduction number are
decent. The figure shows the prior distributions (solid lines), the marginal
posterior distribution (histogram) and the true values used in the simulation
(dashed red line).
</p>


<div id="orge547917" class="figure">
<p><img src="../misc/beast2/bd-sim-ex-1/combined-results.png" alt="combined-results.png" width="800px" />
</p>
</div>

<div class="org-src-container">
<pre class="src src-R"><span style="color: #3a81c3;">library</span><span style="color: #3a81c3;">(</span>dplyr<span style="color: #3a81c3;">)</span>
<span style="color: #3a81c3;">library</span><span style="color: #3a81c3;">(</span>ggplot2<span style="color: #3a81c3;">)</span>
<span style="color: #3a81c3;">library</span><span style="color: #3a81c3;">(</span>magrittr<span style="color: #3a81c3;">)</span>
<span style="color: #3a81c3;">library</span><span style="color: #3a81c3;">(</span>purrr<span style="color: #3a81c3;">)</span>
<span style="color: #3a81c3;">library</span><span style="color: #3a81c3;">(</span>cowplot<span style="color: #3a81c3;">)</span>

birth_rate <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> <span style="color: #4e3163;">104</span>
death_rate <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> <span style="color: #4e3163;">5148</span> / <span style="color: #4e3163;">100</span>
sampling_rate <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> <span style="color: #4e3163;">52</span> / <span style="color: #4e3163;">100</span>
sim_duration <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> <span style="color: #4e3163;">0.15</span> <span style="color: #2aa1ae; background-color: #ecf3ec;"># </span><span style="color: #2aa1ae; background-color: #ecf3ec;">year</span>

mcmc_samples <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> read.csv<span style="color: #3a81c3;">(</span>
  <span style="color: #2d9574;">"sequences.log"</span>,
  comment.char = <span style="color: #2d9574;">"#"</span>,
  sep = <span style="color: #2d9574;">"\t"</span>
<span style="color: #3a81c3;">)</span> |&gt;
  rename<span style="color: #3a81c3;">(</span>
    clock_rate = clockRate,
    origin = origin_BDSKY_Serial,
    net_removal_rate = becomeUninfectiousRate_BDSKY_Serial,
    r_naught = reproductiveNumber_BDSKY_Serial,
    sampling_proportion = samplingProportion_BDSKY_Serial
  <span style="color: #3a81c3;">)</span> |&gt;
  select<span style="color: #3a81c3;">(</span>
    clock_rate,
    origin,
    net_removal_rate,
    r_naught,
    sampling_proportion
  <span style="color: #3a81c3;">)</span> |&gt;
  tail<span style="color: #3a81c3;">(</span>-<span style="color: #4e3163;">1000</span><span style="color: #3a81c3;">)</span>


g1 <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> ggplot<span style="color: #3a81c3;">()</span> +
  geom_histogram<span style="color: #3a81c3;">(</span>
    data = mcmc_samples,
    mapping = aes<span style="color: #6c3163;">(</span>x = r_naught, y = ..density..<span style="color: #6c3163;">)</span>,
    bins = <span style="color: #4e3163;">40</span>
  <span style="color: #3a81c3;">)</span> +
  stat_function<span style="color: #3a81c3;">(</span>
    fun = <span style="color: #3a81c3;">function</span><span style="color: #6c3163;">(</span>x<span style="color: #6c3163;">)</span> dlnorm<span style="color: #6c3163;">(</span>x, meanlog = <span style="color: #4e3163;">0.0</span>, sdlog = <span style="color: #4e3163;">1.0</span><span style="color: #6c3163;">)</span>,
    geom = <span style="color: #2d9574;">"line"</span>
  <span style="color: #3a81c3;">)</span> +
  geom_vline<span style="color: #3a81c3;">(</span>
    xintercept = birth_rate / <span style="color: #6c3163;">(</span>death_rate + sampling_rate<span style="color: #6c3163;">)</span>,
    linetype = <span style="color: #2d9574;">"dashed"</span>,
    colour = <span style="color: #2d9574;">"red"</span>
  <span style="color: #3a81c3;">)</span> +
  xlim<span style="color: #3a81c3;">(</span>c<span style="color: #6c3163;">(</span><span style="color: #4e3163;">0.1</span>, <span style="color: #4e3163;">3</span><span style="color: #6c3163;">)</span><span style="color: #3a81c3;">)</span> +
  labs<span style="color: #3a81c3;">(</span>x = <span style="color: #2d9574;">"R-naught"</span>, y = <span style="color: #2d9574;">"Density"</span><span style="color: #3a81c3;">)</span> +
  theme_classic<span style="color: #3a81c3;">()</span>

g2 <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> ggplot<span style="color: #3a81c3;">()</span> +
  geom_histogram<span style="color: #3a81c3;">(</span>
    data = mcmc_samples,
    mapping = aes<span style="color: #6c3163;">(</span>
      x = net_removal_rate,
      y = ..density..
    <span style="color: #6c3163;">)</span>
  <span style="color: #3a81c3;">)</span> +
  stat_function<span style="color: #3a81c3;">(</span>
    fun = <span style="color: #3a81c3;">function</span><span style="color: #6c3163;">(</span>x<span style="color: #6c3163;">)</span> dnorm<span style="color: #6c3163;">(</span>x, mean = <span style="color: #4e3163;">52</span>, sd = <span style="color: #4e3163;">1.0</span><span style="color: #6c3163;">)</span>,
    geom = <span style="color: #2d9574;">"line"</span>
  <span style="color: #3a81c3;">)</span> +
  geom_vline<span style="color: #3a81c3;">(</span>
    xintercept = death_rate + sampling_rate,
    linetype = <span style="color: #2d9574;">"dashed"</span>,
    colour = <span style="color: #2d9574;">"red"</span>
  <span style="color: #3a81c3;">)</span> +
  xlim<span style="color: #3a81c3;">(</span>c<span style="color: #6c3163;">(</span><span style="color: #4e3163;">48</span>, <span style="color: #4e3163;">56</span><span style="color: #6c3163;">)</span><span style="color: #3a81c3;">)</span> +
  labs<span style="color: #3a81c3;">(</span>x = <span style="color: #2d9574;">"Net removal rate"</span>, y = <span style="color: #2d9574;">"Density"</span><span style="color: #3a81c3;">)</span> +
  theme_classic<span style="color: #3a81c3;">()</span>

g3 <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> ggplot<span style="color: #3a81c3;">()</span> +
  geom_histogram<span style="color: #3a81c3;">(</span>
    data = mcmc_samples,
    mapping = aes<span style="color: #6c3163;">(</span>x = sampling_proportion, y = ..density..<span style="color: #6c3163;">)</span>,
    bins = <span style="color: #4e3163;">20</span>
  <span style="color: #3a81c3;">)</span> +
  stat_function<span style="color: #3a81c3;">(</span>
    fun = <span style="color: #3a81c3;">function</span><span style="color: #6c3163;">(</span>x<span style="color: #6c3163;">)</span> dbeta<span style="color: #6c3163;">(</span>x, shape1 = <span style="color: #4e3163;">2.0</span>, shape2 = <span style="color: #4e3163;">97.0</span><span style="color: #6c3163;">)</span>,
    geom = <span style="color: #2d9574;">"line"</span>
  <span style="color: #3a81c3;">)</span> +
  geom_vline<span style="color: #3a81c3;">(</span>
    xintercept = sampling_rate / <span style="color: #6c3163;">(</span>death_rate + sampling_rate<span style="color: #6c3163;">)</span>,
    linetype = <span style="color: #2d9574;">"dashed"</span>,
    colour = <span style="color: #2d9574;">"red"</span>
  <span style="color: #3a81c3;">)</span> +
  xlim<span style="color: #3a81c3;">(</span>c<span style="color: #6c3163;">(</span><span style="color: #4e3163;">0</span>, <span style="color: #4e3163;">0.1</span><span style="color: #6c3163;">)</span><span style="color: #3a81c3;">)</span> +
  labs<span style="color: #3a81c3;">(</span>x = <span style="color: #2d9574;">"Sampling proportion"</span>, y = <span style="color: #2d9574;">"Density"</span><span style="color: #3a81c3;">)</span> +
  theme_classic<span style="color: #3a81c3;">()</span>

g4 <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> ggplot<span style="color: #3a81c3;">()</span> +
  geom_histogram<span style="color: #3a81c3;">(</span>
    data = mcmc_samples,
    mapping = aes<span style="color: #6c3163;">(</span>x = origin, y = ..density..<span style="color: #6c3163;">)</span>,
    bins = <span style="color: #4e3163;">20</span>
  <span style="color: #3a81c3;">)</span> +
  stat_function<span style="color: #3a81c3;">(</span>
    fun = <span style="color: #3a81c3;">function</span><span style="color: #6c3163;">(</span>x<span style="color: #6c3163;">)</span> dlnorm<span style="color: #6c3163;">(</span>x, meanlog = log<span style="color: #2d9574;">(</span><span style="color: #4e3163;">0.1</span><span style="color: #2d9574;">)</span>, sdlog = <span style="color: #4e3163;">1.25</span><span style="color: #6c3163;">)</span>,
    geom = <span style="color: #2d9574;">"line"</span>
  <span style="color: #3a81c3;">)</span> +
  geom_vline<span style="color: #3a81c3;">(</span>
    xintercept = sim_duration,
    linetype = <span style="color: #2d9574;">"dashed"</span>,
    colour = <span style="color: #2d9574;">"red"</span>
  <span style="color: #3a81c3;">)</span> +
  xlim<span style="color: #3a81c3;">(</span>c<span style="color: #6c3163;">(</span><span style="color: #4e3163;">0</span>, <span style="color: #4e3163;">0.3</span><span style="color: #6c3163;">)</span><span style="color: #3a81c3;">)</span> +
  labs<span style="color: #3a81c3;">(</span>x = <span style="color: #2d9574;">"Origin"</span>, y = <span style="color: #2d9574;">"Density"</span><span style="color: #3a81c3;">)</span> +
  theme_classic<span style="color: #3a81c3;">()</span>

g5 <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> ggplot<span style="color: #3a81c3;">()</span> +
  geom_histogram<span style="color: #3a81c3;">(</span>
    data = mcmc_samples,
    mapping = aes<span style="color: #6c3163;">(</span>x = clock_rate, y = ..density..<span style="color: #6c3163;">)</span>,
    bins = <span style="color: #4e3163;">40</span>
  <span style="color: #3a81c3;">)</span> +
  stat_function<span style="color: #3a81c3;">(</span>
    fun = <span style="color: #3a81c3;">function</span><span style="color: #6c3163;">(</span>x<span style="color: #6c3163;">)</span> dgamma<span style="color: #6c3163;">(</span>x, <span style="color: #4e3163;">10</span>, <span style="color: #4e3163;">1e4</span><span style="color: #6c3163;">)</span>,
    geom = <span style="color: #2d9574;">"line"</span>
  <span style="color: #3a81c3;">)</span> +
  geom_vline<span style="color: #3a81c3;">(</span>
    xintercept = <span style="color: #4e3163;">1e</span>-<span style="color: #4e3163;">3</span>,
    linetype = <span style="color: #2d9574;">"dashed"</span>,
    colour = <span style="color: #2d9574;">"red"</span>
  <span style="color: #3a81c3;">)</span> +
  xlim<span style="color: #3a81c3;">(</span>c<span style="color: #6c3163;">(</span><span style="color: #4e3163;">1e</span>-<span style="color: #4e3163;">5</span>, <span style="color: #4e3163;">3e</span>-<span style="color: #4e3163;">3</span><span style="color: #6c3163;">)</span><span style="color: #3a81c3;">)</span> +
  labs<span style="color: #3a81c3;">(</span>x = <span style="color: #2d9574;">"Clock rate"</span>, y = <span style="color: #2d9574;">"Density"</span><span style="color: #3a81c3;">)</span> +
  theme_classic<span style="color: #3a81c3;">()</span>

g_comb <span style="color: #ba2f59; font-weight: bold;">&lt;-</span> plot_grid<span style="color: #3a81c3;">(</span>
  plot_grid<span style="color: #6c3163;">(</span>g1, g2, g3, nrow = <span style="color: #4e3163;">1</span>, labels = LETTERS<span style="color: #2d9574;">[</span><span style="color: #4e3163;">1</span>:<span style="color: #4e3163;">3</span><span style="color: #2d9574;">]</span><span style="color: #6c3163;">)</span>,
  plot_grid<span style="color: #6c3163;">(</span>g4, g5, nrow = <span style="color: #4e3163;">1</span>, labels = LETTERS<span style="color: #2d9574;">[</span><span style="color: #4e3163;">4</span>:<span style="color: #4e3163;">5</span><span style="color: #2d9574;">]</span><span style="color: #6c3163;">)</span>,
  nrow = <span style="color: #4e3163;">2</span>
<span style="color: #3a81c3;">)</span>

ggsave<span style="color: #3a81c3;">(</span>
  filename = <span style="color: #2d9574;">"combined-results.png"</span>,
  plot = g_comb,
  height = <span style="color: #4e3163;">14.8</span>, width = <span style="color: #4e3163;">21.0</span>,
  units = <span style="color: #2d9574;">"cm"</span>
<span style="color: #3a81c3;">)</span>
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-org3af6932" class="outline-3">
<h3 id="org3af6932">Birth-death simulation: Example II</h3>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Alex Zarebski</p>
<p class="date">Created: 2021-12-23 Thu 15:20</p>
<p class="validation"><a href="https://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
