<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<!-- 2024-04-12 Fri 10:29 -->
<meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<title>phylogenetics</title>
<meta name="author" content="Alexander E. Zarebski" />
<meta name="generator" content="Org Mode" />
<style>
  #content { max-width: 60em; margin: auto; }
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #e6e6e6;
    border-radius: 3px;
    background-color: #f2f2f2;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: auto;
  }
  pre.src:before {
    display: none;
    position: absolute;
    top: -8px;
    right: 12px;
    padding: 3px;
    color: #555;
    background-color: #f2f2f299;
  }
  pre.src:hover:before { display: inline; margin-top: 14px;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-authinfo::before { content: 'Authinfo'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .equation-container {
    display: table;
    text-align: center;
    width: 100%;
  }
  .equation {
    vertical-align: middle;
  }
  .equation-label {
    display: table-cell;
    text-align: right;
    vertical-align: middle;
  }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { }
</style>
<link rel="icon" type="image/png" href="../resources/nicemacs-favicon.png">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/milligram/1.4.1/milligram.css">
<link rel="stylesheet" href="../microgram.css">
<script>
  window.MathJax = {
    tex: {
      ams: {
        multlineWidth: '85%'
      },
      tags: 'ams',
      tagSide: 'right',
      tagIndent: '.8em'
    },
    chtml: {
      scale: 1.0,
      displayAlign: 'center',
      displayIndent: '0em'
    },
    svg: {
      scale: 1.0,
      displayAlign: 'center',
      displayIndent: '0em'
    },
    output: {
      font: 'mathjax-modern',
      displayOverflow: 'overflow'
    }
  };
</script>

<script
  id="MathJax-script"
  async
  src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js">
</script>
</head>
<body>
<div id="content" class="content">
<h1 class="title">phylogenetics</h1>
<div id="table-of-contents" role="doc-toc">
<h2>Table of Contents</h2>
<div id="text-table-of-contents" role="doc-toc">
<ul>
<li><a href="#orgf1ef9d0">Grammars</a>
<ul>
<li><a href="#orgb02b54c">Newick</a></li>
</ul>
</li>
<li><a href="#org3da4495">Software</a>
<ul>
<li><a href="#org6f5d5cc">biopython</a>
<ul>
<li><a href="#org54425f0">Example: Reading Newick from file</a></li>
<li><a href="#org667c4e1">Example: Reading Newick from string</a></li>
<li><a href="#orgc0dc4eb">Example: Plotting a tree</a></li>
<li><a href="#orgd3fbbd0">Example: Counting the number of leaves</a></li>
<li><a href="#org0d2b318">Example: Normalising a tree to unit branch lengths</a></li>
<li><a href="#orgca43778">Example: Quickly reading a tree from disk</a></li>
</ul>
</li>
<li><a href="#orgaf70897">Beast2</a></li>
<li><a href="#org6e95efc">TempEst</a>
<ul>
<li><a href="#orged25e30">Example 1</a></li>
<li><a href="#orgdfb08fc">Example 2</a></li>
</ul>
</li>
<li><a href="#orgbd85e88">Rentrez in R (downloading sequence data)</a></li>
<li><a href="#orgca5b21b">APE in R</a>
<ul>
<li><a href="#org132680e">The <code>phylo</code> class</a></li>
<li><a href="#org2ca641d">Simulation</a></li>
</ul>
</li>
<li><a href="#orgfaa80be">Phangorn in R</a></li>
<li><a href="#org66e86f0">Ggtree and friends</a></li>
</ul>
</li>
</ul>
</div>
</div>
<p>
<a href="../index.html">Home</a>
</p>

<div id="outline-container-orgf1ef9d0" class="outline-2">
<h2 id="orgf1ef9d0">Grammars</h2>
<div class="outline-text-2" id="text-orgf1ef9d0">
</div>
<div id="outline-container-orgb02b54c" class="outline-3">
<h3 id="orgb02b54c">Newick</h3>
<div class="outline-text-3" id="text-orgb02b54c">
<p>
See <a href="https://web.archive.org/web/20210225204509/https://evolution.genetics.washington.edu/phylip/newick_doc.html">here</a> for a formal specification of the Newick grammar by Gary Olsen (1990).
The important part of this is
</p>

<div class="org-src-container">
<pre class="src src-text">Conventions:
   Items in { } may appear zero or more times.
   Items in [ ] are optional, they may appear once or not at all.
   All other punctuation marks (colon, semicolon, parentheses, comma and
         single quote) are required parts of the format.


              tree ==&gt; descendant_list [ root_label ] [ : branch_length ] ;

   descendant_list ==&gt; ( subtree { , subtree } )

           subtree ==&gt; descendant_list [internal_node_label] [: branch_length]
                   ==&gt; leaf_label [: branch_length]

            root_label ==&gt; label
   internal_node_label ==&gt; label
            leaf_label ==&gt; label

                 label ==&gt; unquoted_label
                       ==&gt; quoted_label

        unquoted_label ==&gt; string_of_printing_characters
          quoted_label ==&gt; ' string_of_printing_characters '

         branch_length ==&gt; signed_number
                       ==&gt; unsigned_number
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-org3da4495" class="outline-2">
<h2 id="org3da4495">Software</h2>
<div class="outline-text-2" id="text-org3da4495">
</div>
<div id="outline-container-org6f5d5cc" class="outline-3">
<h3 id="org6f5d5cc">biopython</h3>
<div class="outline-text-3" id="text-org6f5d5cc">

<div id="org10a72b4" class="figure">
<p><img src="../images/biopython-logo.png" alt="biopython-logo.png" width="300px" />
</p>
</div>

<p>
<a href="https://biopython.org/docs/1.75/api/index.html">Biopython</a> is a Python library that aims to provide a comprehensive set
of modules for computational biology and bioinformatics.
</p>

<ul class="org-ul">
<li>the <a href="https://biopython.org/docs/1.75/api/Bio.Phylo.BaseTree.html">Bio Phylo BaseTree</a> module provides foundational classes for
working with phylogenies.</li>
<li>the <a href="https://biopython.org/docs/1.75/api/Bio.Phylo.BaseTree.html#Bio.Phylo.BaseTree.Clade">Clade</a> class in particular provides a recursively defined tree.
<ul class="org-ul">
<li>instances of Clade have <code>branch_length</code> (the length of the branch
to their parent) and <code>clades</code> (a list of their children).</li>
<li>the Clade class has a method <code>is_terminal</code> which is a predicate
for if the current clade is a leaf node.</li>
<li>iterating over a Clade will iterate over its direct children.</li>
<li>the <code>__len__</code> of a Clade is its number of children.</li>
</ul></li>
</ul>
</div>

<div id="outline-container-org54425f0" class="outline-4">
<h4 id="org54425f0">Example: Reading Newick from file</h4>
<div class="outline-text-4" id="text-org54425f0">
<div class="org-src-container">
<pre class="src src-python"><span style="color: #0000FF;">from</span> Bio <span style="color: #0000FF;">import</span> Phylo

<span style="color: #BA36A5;">newick_tree</span> = <span style="color: #008000;">"((A:0.1, B:0.2):0.3, (C:0.4, D:0.5):0.6);"</span>

<span style="color: #BA36A5;">filename</span> = <span style="color: #008000;">"foobar.newick"</span>
<span style="color: #0000FF;">with</span> <span style="color: #006FE0;">open</span>(filename, <span style="color: #008000;">"w+"</span>) <span style="color: #0000FF;">as</span> <span style="color: #006FE0;">file</span>:
    <span style="color: #006FE0;">file</span>.write(newick_tree)

<span style="color: #BA36A5;">tree</span> = Phylo.read(filename, <span style="color: #008000;">"newick"</span>)
<span style="color: #BA36A5;">clade</span> = tree.root

<span style="color: #0000FF;">assert</span> clade[0][0].branch_length == 0.1
<span style="color: #0000FF;">assert</span> clade[0][0].name == <span style="color: #008000;">"A"</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-org667c4e1" class="outline-4">
<h4 id="org667c4e1">Example: Reading Newick from string</h4>
<div class="outline-text-4" id="text-org667c4e1">
<div class="org-src-container">
<pre class="src src-python"><span style="color: #0000FF;">from</span> io <span style="color: #0000FF;">import</span> StringIO
<span style="color: #0000FF;">from</span> Bio <span style="color: #0000FF;">import</span> Phylo

<span style="color: #BA36A5;">newick_tree</span> = StringIO(<span style="color: #008000;">"((A:0.1, B:0.2):0.3, (C:0.4, D:0.5):0.6);"</span>)

<span style="color: #BA36A5;">tree</span> = Phylo.read(newick_tree, <span style="color: #008000;">"newick"</span>)
<span style="color: #BA36A5;">clade</span> = tree.root

<span style="color: #0000FF;">assert</span> clade[0][0].branch_length == 0.1
<span style="color: #0000FF;">assert</span> clade[0][0].name == <span style="color: #008000;">"A"</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-orgc0dc4eb" class="outline-4">
<h4 id="orgc0dc4eb">Example: Plotting a tree</h4>
<div class="outline-text-4" id="text-orgc0dc4eb">

<div id="orgd35882f" class="figure">
<p><img src="../images/biopython-phylo-draw.png" alt="biopython-phylo-draw.png" width="400px" />
</p>
<p><span class="figure-number">Figure 1: </span>Plot generated by the <code>Phylo.draw</code> method.</p>
</div>

<div class="org-src-container">
<pre class="src src-python"><span style="color: #0000FF;">from</span> io <span style="color: #0000FF;">import</span> StringIO
<span style="color: #0000FF;">from</span> Bio <span style="color: #0000FF;">import</span> Phylo

<span style="color: #BA36A5;">newick_tree</span> = StringIO(<span style="color: #008000;">"((A:0.1, B:0.2):0.3, (C:0.4, D:0.5):0.6);"</span>)

<span style="color: #BA36A5;">tree</span> = Phylo.read(newick_tree, <span style="color: #008000;">"newick"</span>)
Phylo.draw(tree)
</pre>
</div>
</div>
</div>

<div id="outline-container-orgd3fbbd0" class="outline-4">
<h4 id="orgd3fbbd0">Example: Counting the number of leaves</h4>
<div class="outline-text-4" id="text-orgd3fbbd0">
<div class="org-src-container">
<pre class="src src-python"><span style="color: #0000FF;">from</span> io <span style="color: #0000FF;">import</span> StringIO
<span style="color: #0000FF;">from</span> Bio <span style="color: #0000FF;">import</span> Phylo

<span style="color: #BA36A5;">newick_tree</span> = StringIO(<span style="color: #008000;">"((A:0.1, B:0.2):0.3, (C:0.4, D:0.5):0.6);"</span>)

<span style="color: #BA36A5;">tree</span> = Phylo.read(newick_tree, <span style="color: #008000;">"newick"</span>)
<span style="color: #006FE0;">print</span>(f<span style="color: #008000;">"There are </span>{<span style="color: #006FE0;">len</span>(tree.get_terminals())}<span style="color: #008000;"> leaves"</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-example">There are 4 leaves
</pre>
</div>
</div>
</div>

<div id="outline-container-org0d2b318" class="outline-4">
<h4 id="org0d2b318">Example: Normalising a tree to unit branch lengths</h4>
<div class="outline-text-4" id="text-org0d2b318">
<div class="org-src-container">
<pre class="src src-python"><span style="color: #0000FF;">from</span> io <span style="color: #0000FF;">import</span> StringIO
<span style="color: #0000FF;">from</span> Bio <span style="color: #0000FF;">import</span> Phylo
<span style="color: #0000FF;">import</span> copy

<span style="color: #BA36A5;">newick_tree</span> = StringIO(<span style="color: #008000;">"((A:0.1, B:0.2):0.3, (C:0.4, D:0.5):0.6);"</span>)

<span style="color: #BA36A5;">tree</span> = Phylo.read(newick_tree, <span style="color: #008000;">"newick"</span>)
<span style="color: #0000FF;">assert</span> <span style="color: #006FE0;">max</span>(tree.depths().values()) == 1.1

<span style="color: #0000FF;">def</span> <span style="color: #006699;">normalise_branch_lengths</span>(tree):
    <span style="color: #036A07;">"""</span>
<span style="color: #036A07;">    Normalise a tree to unit branch lengths.</span>
<span style="color: #036A07;">    """</span>
    <span style="color: #BA36A5;">new_tree</span> = copy.deepcopy(tree)
    <span style="color: #BA36A5;">branch_lens</span> = [c.branch_length <span style="color: #0000FF;">for</span> c <span style="color: #0000FF;">in</span> new_tree.find_clades()
                   <span style="color: #0000FF;">if</span> c.branch_length <span style="color: #0000FF;">is</span> <span style="color: #0000FF;">not</span> <span style="color: #D0372D;">None</span>]
    <span style="color: #BA36A5;">mean_branch_len</span> = <span style="color: #006FE0;">sum</span>(branch_lens) / <span style="color: #006FE0;">len</span>(branch_lens)
    <span style="color: #BA36A5;">scale_factor</span> = 1 / mean_branch_len
    <span style="color: #0000FF;">for</span> c <span style="color: #0000FF;">in</span> new_tree.find_clades():
        <span style="color: #0000FF;">if</span> c.branch_length <span style="color: #0000FF;">is</span> <span style="color: #0000FF;">not</span> <span style="color: #D0372D;">None</span>:
            c.<span style="color: #BA36A5;">branch_length</span> *= scale_factor
    <span style="color: #0000FF;">return</span> mean_branch_len, new_tree

<span style="color: #BA36A5;">mean_branch_len</span>, <span style="color: #BA36A5;">new_tree</span> = normalise_branch_lengths(tree)

<span style="color: #006FE0;">print</span>(f<span style="color: #008000;">"The mean branch length was </span>{<span style="color: #006FE0;">format</span>(mean_branch_len, '.3f')}<span style="color: #008000;">"</span>)
<span style="color: #006FE0;">print</span>(f<span style="color: #008000;">"The old tree height is </span>{<span style="color: #006FE0;">max</span>(tree.depths().values())}<span style="color: #008000;">"</span>)
<span style="color: #006FE0;">print</span>(f<span style="color: #008000;">"The new tree height is </span>{<span style="color: #006FE0;">format</span>(<span style="color: #006FE0;">max</span>(new_tree.depths().values()), '.3f')}<span style="color: #008000;">"</span>)
<span style="color: #0000FF;">assert</span> <span style="color: #006FE0;">max</span>(tree.depths().values()) == 1.1
<span style="color: #0000FF;">assert</span> <span style="color: #006FE0;">abs</span>(<span style="color: #006FE0;">max</span>(new_tree.depths().values()) - (1.1 * scale_factor)) &lt; 1e-6
</pre>
</div>

<p>
The output of this script is
</p>

<pre class="example" id="org6ee35a5">
The mean branch length was 0.350
The old tree height is 1.1
The new tree height is 3.143
</pre>
</div>
</div>

<div id="outline-container-orgca43778" class="outline-4">
<h4 id="orgca43778">Example: Quickly reading a tree from disk</h4>
<div class="outline-text-4" id="text-orgca43778">
<p>
<b>TLDR:</b> <code>pickle</code> is faster than Newick.
</p>


<p>
The following code writes a big tree to one of two types of files: a
plain text Newick file or a pickle. It then reads those trees back
multiple times to get an estimate of how long it takes. Across a range
of tests we see that the pickled version is about twice as fast, but
this seems to diminish as we get to larger trees.
</p>

<div class="org-src-container">
<pre class="src src-python"><span style="color: #0000FF;">from</span> io <span style="color: #0000FF;">import</span> StringIO
<span style="color: #0000FF;">from</span> Bio <span style="color: #0000FF;">import</span> Phylo
<span style="color: #0000FF;">import</span> pickle
<span style="color: #0000FF;">import</span> timeit

<span style="color: #0000FF;">def</span> <span style="color: #006699;">_tree_string</span>(n):
    <span style="color: #0000FF;">if</span> n == 0:
        <span style="color: #0000FF;">return</span> <span style="color: #008000;">"(A:0.1, A:0.1):0.1"</span>
    <span style="color: #0000FF;">else</span>:
        <span style="color: #BA36A5;">tmp</span> = _tree_string(n-1)
        <span style="color: #0000FF;">return</span> f<span style="color: #008000;">"(</span>{tmp}<span style="color: #008000;">, </span>{tmp}<span style="color: #008000;">):0.1"</span>

<span style="color: #0000FF;">def</span> <span style="color: #006699;">tree_string</span>(n):
    <span style="color: #0000FF;">return</span> _tree_string(n) + <span style="color: #008000;">";"</span>

<span style="color: #BA36A5;">newick_tree</span> = tree_string(9)

<span style="color: #BA36A5;">filename_newick</span> = <span style="color: #008000;">"foobar.newick"</span>
<span style="color: #0000FF;">with</span> <span style="color: #006FE0;">open</span>(filename_newick, <span style="color: #008000;">"w+"</span>) <span style="color: #0000FF;">as</span> <span style="color: #006FE0;">file</span>:
    <span style="color: #006FE0;">file</span>.write(newick_tree)

<span style="color: #BA36A5;">filename_pickle</span> = <span style="color: #008000;">"foobar.pickle"</span>
<span style="color: #BA36A5;">newick_tree_obj</span> = Phylo.read(StringIO(newick_tree), <span style="color: #008000;">"newick"</span>)
<span style="color: #0000FF;">with</span> <span style="color: #006FE0;">open</span>(filename_pickle, <span style="color: #008000;">"wb+"</span>) <span style="color: #0000FF;">as</span> <span style="color: #006FE0;">file</span>:
    pickle.dump(newick_tree_obj, <span style="color: #006FE0;">file</span>)


<span style="color: #0000FF;">def</span> <span style="color: #006699;">test_a</span>():
    <span style="color: #BA36A5;">tree</span> = Phylo.read(filename_newick, <span style="color: #008000;">"newick"</span>)
    <span style="color: #0000FF;">return</span> <span style="color: #006FE0;">len</span>(tree.get_terminals())

<span style="color: #0000FF;">def</span> <span style="color: #006699;">test_b</span>():
    <span style="color: #0000FF;">with</span> <span style="color: #006FE0;">open</span>(filename_pickle, <span style="color: #008000;">"rb"</span>) <span style="color: #0000FF;">as</span> <span style="color: #006FE0;">file</span>:
        <span style="color: #BA36A5;">tree</span> = pickle.load(<span style="color: #006FE0;">file</span>)
        <span style="color: #0000FF;">return</span> <span style="color: #006FE0;">len</span>(tree.get_terminals())

<span style="color: #0000FF;">assert</span> test_a() == test_b()

<span style="color: #BA36A5;">num_reps</span> = 250
<span style="color: #BA36A5;">time_a</span> = timeit.timeit(<span style="color: #0000FF;">lambda</span>: test_a(), number=num_reps)
<span style="color: #BA36A5;">time_b</span> = timeit.timeit(<span style="color: #0000FF;">lambda</span>: test_b(), number=num_reps)
<span style="color: #006FE0;">print</span>(f<span style="color: #008000;">"The ratio of time needed is </span>{(time_a/time_b):.2f}<span style="color: #008000;"> for Bio/pickle with </span>{test_a()}<span style="color: #008000;"> nodes with </span>{num_reps}<span style="color: #008000;"> replicates."</span>)
</pre>
</div>

<p>
Gives us the following examples for different tree sizes.
</p>

<pre class="example" id="orge66f70d">
The ratio of time needed is 1.98 for Bio/pickle with 256 nodes with 250 replicates.
The ratio of time needed is 2.00 for Bio/pickle with 512 nodes with 250 replicates.
The ratio of time needed is 1.80 for Bio/pickle with 1024 nodes with 250 replicates.
The ratio of time needed is 1.69 for Bio/pickle with 2048 nodes with 250 replicates.
</pre>
</div>
</div>
</div>

<div id="outline-container-orgaf70897" class="outline-3">
<h3 id="orgaf70897">Beast2</h3>
<div class="outline-text-3" id="text-orgaf70897">
<p>
<a href="./beast2-notes.html">My BEAST2 notes</a>
</p>
</div>
</div>

<div id="outline-container-org6e95efc" class="outline-3">
<h3 id="org6e95efc">TempEst</h3>
<div class="outline-text-3" id="text-org6e95efc">
<ul class="org-ul">
<li><a href="https://beast.community/tempest">BEAST page on TempEst</a></li>
<li><a href="http://tree.bio.ed.ac.uk/software/tempest/">Download from here</a></li>
<li><a href="https://doi.org/10.1093/ve/vew007">DOI: 10.1093/ve/vew007</a></li>
</ul>
</div>

<div id="outline-container-orged25e30" class="outline-4">
<h4 id="orged25e30">Example 1</h4>
<div class="outline-text-4" id="text-orged25e30">
<p>
The following examples demonstrates how to use <code>ape::dist.dna</code> and
<code>ape::nj</code> to get a simple tree that can be read into TempEst.
</p>

<div class="org-src-container">
<pre class="src src-R">alignment_dna <span style="color: #D0372D;">&lt;-</span> read.dna(<span style="color: #008000;">"alignment.fasta"</span>, format = <span style="color: #008000;">"fasta"</span>)
dist_matrix <span style="color: #D0372D;">&lt;-</span> dist.dna(alignment_dna, model = <span style="color: #008000;">"K80"</span>, pairwise.deletion = <span style="color: #6434A3;">TRUE</span>)
nj_phylo <span style="color: #D0372D;">&lt;-</span> nj(dist_matrix)
write.nexus(nj_phylo, file = <span style="color: #008000;">"nj-tree.nexus"</span>)
</pre>
</div>
</div>
</div>

<div id="outline-container-orgdfb08fc" class="outline-4">
<h4 id="orgdfb08fc">Example 2</h4>
<div class="outline-text-4" id="text-orgdfb08fc">
<p>
We will use the tree shown in Figure <a href="#orgd6c94eb">1</a> an example
</p>


<div id="orgd6c94eb" class="figure">
<p><img src="../images/tempest-demo-tree.png" alt="tempest-demo-tree.png" width="300px" />
</p>
</div>

<p>
When we simulate some sequences on this and then create a neighbour-joining tree
based on these sequences we can put this into TempEst to check if there is any
signal in the data. The screen shot in Figure <a href="#orga9e5c09">2</a> shows the results. Note that
even with ideal data the result is not perfect.
</p>


<div id="orga9e5c09" class="figure">
<p><img src="../images/tempest-results.png" alt="tempest-results.png" width="500px" />
</p>
</div>

<p>
Here is the code that simulated the data set.
</p>

<div class="org-src-container">
<pre class="src src-R"><span style="color: #D0372D;">library</span>(ape)
<span style="color: #D0372D;">library</span>(phangorn)
set.seed(1)
<span style="color: #000000; font-weight: bold;">#</span><span style="color: #000000; font-weight: bold; font-style: italic;">'</span><span style="color: #8D8D84; font-style: italic;"> Read in a simple tree to use as an example so we do not</span>
<span style="color: #000000; font-weight: bold;">#</span><span style="color: #000000; font-weight: bold; font-style: italic;">'</span><span style="color: #8D8D84; font-style: italic;"> need to mess around trying to simulate one.</span>
demo_tree <span style="color: #D0372D;">&lt;-</span> read.tree(text = <span style="color: #008000;">"(((((A:2.0,B:1.0):2.0,C:1.0):2.0,D:1.0):2.0,E:1.0):2.0,F:1.0);"</span>)
png(filename=<span style="color: #008000;">"tempest-demo-tree.png"</span>)
plot(demo_tree)
dev.off()
num_tips <span style="color: #D0372D;">&lt;-</span> length(demo_tree$tip.label)
tip_times <span style="color: #D0372D;">&lt;-</span> head(
  node.depth.edgelength(demo_tree),
  num_tips
)
<span style="color: #000000; font-weight: bold;">#</span><span style="color: #000000; font-weight: bold; font-style: italic;">'</span><span style="color: #8D8D84; font-style: italic;"> Relabel the tips to preserve the sampling date in a</span>
<span style="color: #000000; font-weight: bold;">#</span><span style="color: #000000; font-weight: bold; font-style: italic;">'</span><span style="color: #8D8D84; font-style: italic;"> suitable format for subsequent use.</span>
demo_tree$tip.label <span style="color: #D0372D;">&lt;-</span> sprintf(<span style="color: #008000;">"%s_%f"</span>, demo_tree$tip.label, tip_times)
demo_seqs <span style="color: #D0372D;">&lt;-</span> demo_tree |&gt; simSeq(l = 1e4, rate = 1e-3) |&gt; as.character()
<span style="color: #000000; font-weight: bold;">#</span><span style="color: #000000; font-weight: bold; font-style: italic;">'</span><span style="color: #8D8D84; font-style: italic;"> Compute the distances manually because the functions</span>
<span style="color: #000000; font-weight: bold;">#</span><span style="color: #000000; font-weight: bold; font-style: italic;">'</span><span style="color: #8D8D84; font-style: italic;"> provided seem to be a bit buggy.</span>
demo_distances <span style="color: #D0372D;">&lt;-</span> matrix(0, num_tips, num_tips)
rownames(demo_distances) <span style="color: #D0372D;">&lt;-</span> demo_tree$tip.label
colnames(demo_distances) <span style="color: #D0372D;">&lt;-</span> demo_tree$tip.label
<span style="color: #0000FF;">for</span> (ix <span style="color: #0000FF;">in</span> 1:(num_tips - 1)) {
  <span style="color: #0000FF;">for</span> (jx <span style="color: #0000FF;">in</span> ix:num_tips) {
    jac <span style="color: #D0372D;">&lt;-</span> sum(demo_seqs[ix,] != demo_seqs[jx,])
    demo_distances[ix,jx] <span style="color: #D0372D;">&lt;-</span> jac
    demo_distances[jx,ix] <span style="color: #D0372D;">&lt;-</span> jac
  }
}
<span style="color: #000000; font-weight: bold;">#</span><span style="color: #000000; font-weight: bold; font-style: italic;">'</span><span style="color: #8D8D84; font-style: italic;"> Estimate the tree using the distances and write this to a</span>
<span style="color: #000000; font-weight: bold;">#</span><span style="color: #000000; font-weight: bold; font-style: italic;">'</span><span style="color: #8D8D84; font-style: italic;"> nexus file so it can be read into TempEst.</span>
demo_nj <span style="color: #D0372D;">&lt;-</span> nj(demo_distances)
write.nexus(demo_nj, file = <span style="color: #008000;">"tempest-demo-data.nexus"</span>)
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-orgbd85e88" class="outline-3">
<h3 id="orgbd85e88">Rentrez in R (downloading sequence data)</h3>
<div class="outline-text-3" id="text-orgbd85e88">
<p>
By default, <code>rentrez</code> returns fetched data as an <code>XML</code> object from the
XML package, rather than as an <code>xml_node</code> from the <code>xml2</code> package,
which is much easier to work with, so I'll add a helper here to move
between them.
</p>

<div class="org-src-container">
<pre class="src src-R"><span style="color: #000000; font-weight: bold;">#</span><span style="color: #000000; font-weight: bold; font-style: italic;">'</span><span style="color: #8D8D84; font-style: italic;"> Fetch Entrez data as XML object</span>
<span style="color: #000000; font-weight: bold;">#</span><span style="color: #000000; font-weight: bold; font-style: italic;">'</span>
<span style="color: #000000; font-weight: bold;">#</span><span style="color: #000000; font-weight: bold; font-style: italic;">'</span><span style="color: #8D8D84; font-style: italic;"> </span><span style="color: #0000FF; font-style: italic;">@param</span><span style="color: #8D8D84; font-style: italic;"> </span><span style="color: #BA36A5; font-style: italic;">...</span><span style="color: #8D8D84; font-style: italic;"> Arguments to pass to =rentrez::entrez_fetch=</span>
<span style="color: #000000; font-weight: bold;">#</span><span style="color: #000000; font-weight: bold; font-style: italic;">'</span>
<span style="color: #000000; font-weight: bold;">#</span><span style="color: #000000; font-weight: bold; font-style: italic;">'</span><span style="color: #8D8D84; font-style: italic;"> </span><span style="color: #0000FF; font-style: italic;">@return</span><span style="color: #8D8D84; font-style: italic;"> An =xml_node= object</span>
<span style="color: #000000; font-weight: bold;">#</span><span style="color: #000000; font-weight: bold; font-style: italic;">'</span>
<span style="color: #000000; font-weight: bold;">#</span><span style="color: #000000; font-weight: bold; font-style: italic;">'</span><span style="color: #8D8D84; font-style: italic;"> </span><span style="color: #0000FF; font-style: italic;">@examples</span>
<span style="color: #000000; font-weight: bold;">#</span><span style="color: #000000; font-weight: bold; font-style: italic;">'</span><span style="color: #8D8D84; font-style: italic;"> entrez_obj &lt;- fetch_as_xml(db = "nucleotide", id = accession_numbers)</span>
<span style="color: #000000; font-weight: bold;">#</span><span style="color: #000000; font-weight: bold; font-style: italic;">'</span>
<span style="color: #006699;">fetch_as_xml</span> <span style="color: #D0372D;">&lt;-</span> <span style="color: #0000FF;">function</span>(...) {
  temp_file <span style="color: #D0372D;">&lt;-</span> tempfile()
  <span style="color: #0000FF;">on.exit</span>(file.remove(temp_file))

  entrez_obj <span style="color: #D0372D;">&lt;-</span> rentrez::entrez_fetch(
    ...,
    rettype = <span style="color: #008000;">"xml"</span>,
    parsed = <span style="color: #6434A3;">TRUE</span>
  )
  XML::saveXML(entrez_obj, file = temp_file)
  <span style="color: #0000FF;">return</span>(xml2::read_xml(temp_file))
}
</pre>
</div>
</div>
</div>

<div id="outline-container-orgca5b21b" class="outline-3">
<h3 id="orgca5b21b">APE in R</h3>
<div class="outline-text-3" id="text-orgca5b21b">
<p>
<a href="https://cran.r-project.org/web/packages/ape/ape.pdf">FULL DOCUMENTATION</a>
</p>

<p>
Use <code>rlineage</code> to simulate the complete tree. To get the sub-tree consisting of
just those that made it to the present use <code>drop.fossils</code>. If you want to filter
for particular nodes, there is <code>keep.tip</code> and <code>drop.tip</code>.
</p>

<p>
There is a <code>plot.phylo</code> method and <code>ltt.plot</code> which can be used to get quick
visualisations of a phylogeny, but for publication quality figures you probably
want to use Ggtree.
</p>
</div>

<div id="outline-container-org132680e" class="outline-4">
<h4 id="org132680e">The <code>phylo</code> class</h4>
<div class="outline-text-4" id="text-org132680e">
<p>
The <code>phylo</code> class is pretty much the defacto standard for representing
phylogenies in R. A phylogeny with \(n\) tips is represented as a <code>phylo</code> which
has the following attributes:
</p>

<ul class="org-ul">
<li><code>Nnode</code> is the number of internal nodes in the tree,</li>
<li><code>tip.label</code> is a string indicating the name of each tip,</li>
<li><code>edge</code> is a matrix with two columns where a row, \([i,j]\), indicates an edge
from node \(i\) to node \(j\),</li>
<li><code>edge.length</code> is an array where element is the length of an edge in the
corresponding row of the <code>edge</code> matrix.</li>
</ul>

<p>
<b>Note</b> that the indexing of the nodes of the tree is \(1\) to \(n\) for the tips
and the internal nodes get allocated the integers \(n+1,n+2,\dots\). You do not
want to reconstruct the tree from this information yourself though so <code>ape</code>
offers convenience functions such as <code>node.depth.edgelength</code> which gives you the
depths from the TMRCA down to each node.
</p>
</div>
</div>

<div id="outline-container-org2ca641d" class="outline-4">
<h4 id="org2ca641d">Simulation</h4>
<div class="outline-text-4" id="text-org2ca641d">
<p>
I wrote a little tool that uses ape to simulate from the
birth-death-sampling-occurrence process. It is available in this <a href="https://gist.github.com/aezarebski/346684e01831be9293ef1830350a4d97">gist</a>.
</p>

<p>
To simulate from the process with time varying rates, here is a little example.
Note that this uses forwards time starting from the origin as time zero. Figure
<a href="#orgfbb7bca">X</a> shows the output of this simulation.
</p>

<div class="org-src-container">
<pre class="src src-R"><span style="color: #D0372D;">library</span>(ape)
<span style="color: #D0372D;">library</span>(purrr)
set.seed(1)

<span style="color: #006699;">birth_rate</span> <span style="color: #D0372D;">&lt;-</span> <span style="color: #0000FF;">function</span>(ts) {
  ifelse(ts &lt; 0.5, 5.0, 1.0)
}

<span style="color: #006699;">death_rate</span> <span style="color: #D0372D;">&lt;-</span> <span style="color: #0000FF;">function</span>(ts) {
  rep(2.0, length(ts))
}

<span style="color: #006699;">max_ltt</span> <span style="color: #D0372D;">&lt;-</span> <span style="color: #0000FF;">function</span>(x) {
  max(ltt.plot.coords(x)[,2])
}

ps <span style="color: #D0372D;">&lt;-</span> purrr::map(1:100, \(x) rlineage(birth_rate, death_rate, Tmax = 1))

y_upper_lim <span style="color: #D0372D;">&lt;-</span> purrr::map(ps, max_ltt) |&gt; as_vector() |&gt; max()
png(<span style="color: #008000;">'ape-rlineage-demo.png'</span>,
    width = 1.618 * 300,
    height = 300,
    units = <span style="color: #008000;">"px"</span>)
ltt.plot(ps[[1]], ylim = c(1, y_upper_lim))
purrr::walk(tail(ps, -1), ltt.lines)
abline(v=-0.5, col=<span style="color: #008000;">'red'</span>)
dev.off()
</pre>
</div>


<div id="orgfbb7bca" class="figure">
<p><img src="../images/ape-rlineage-demo.png" alt="ape-rlineage-demo.png" width="450px" />
</p>
</div>
</div>
</div>
</div>

<div id="outline-container-orgfaa80be" class="outline-3">
<h3 id="orgfaa80be">Phangorn in R</h3>
<div class="outline-text-3" id="text-orgfaa80be">
<p>
The <a href="https://cran.r-project.org/web/packages/phangorn/vignettes/Trees.html">Trees</a> vignette with the phangorn package gives a very clear explanation of
how to estimate a maximum likelihood tree to describe a time stamped MSA. Here
is a summary of the steps involved with estimating a tree for H3N2 sequences
that are bundled with the package.
</p>

<p>
First we load in the time stamped sequences that are bundled with the package
and construct a named vector with the tip dates. Note that older versions of the
package do not come with this dataset so you may need to update to a more recent
version, otherwise you can get it from a recent copy of the source code on
GitHub.
</p>

<div class="org-src-container">
<pre class="src src-R"><span style="color: #D0372D;">library</span>(ape)
<span style="color: #D0372D;">library</span>(phangorn)

fdir <span style="color: #D0372D;">&lt;-</span> system.file(<span style="color: #008000;">"extdata/trees"</span>, package = <span style="color: #008000;">"phangorn"</span>)
tmp <span style="color: #D0372D;">&lt;-</span> read.csv(file.path(fdir, <span style="color: #008000;">"H3N2_NA_20.csv"</span>))
H3N2 <span style="color: #D0372D;">&lt;-</span> read.phyDat(file.path(fdir, <span style="color: #008000;">"H3N2_NA_20.fasta"</span>), format = <span style="color: #008000;">"fasta"</span>)
dates <span style="color: #D0372D;">&lt;-</span> setNames(tmp$numdate_given, tmp$name)
</pre>
</div>

<p>
Here we use the black box function to estimate the tree with a \(\text{GTR} +
\Gamma(4)\) model and dated tips. There are far more details in the the vignette
about how to do this without just calling a black box, but this is a quick way
to get something to start subsequent analyses with.
</p>

<div class="org-src-container">
<pre class="src src-R">fit_td <span style="color: #D0372D;">&lt;-</span> pml_bb(
  H3N2,
  model = <span style="color: #008000;">"GTR+G(4)"</span>,
  method = <span style="color: #008000;">"tipdated"</span>,
  tip.dates = dates,
  rearrangment = <span style="color: #008000;">"NNI"</span>,
  control = pml.control(trace = 0)
)
</pre>
</div>

<p>
Then we can make a simple plot showing the estimated tree.
</p>

<div class="org-src-container">
<pre class="src src-R">tree_td <span style="color: #D0372D;">&lt;-</span> fit_td$tree
root_time <span style="color: #D0372D;">&lt;-</span> max(dates) - max(node.depth.edgelength(tree_td))

png(<span style="color: #008000;">"phangorn-demo.png"</span>,
  width = 1.618 * 300,
  height = 300,
  units = <span style="color: #008000;">"px"</span>
)
plot(tree_td, show.tip.label = <span style="color: #6434A3;">FALSE</span>)
title(main = <span style="color: #008000;">"H3N2"</span>, adj = 1)
title(sub = <span style="color: #008000;">"Timed tree estimated using phangorn"</span>, line = -13, adj = 1)
axisPhylo(root.time = root_time, backward = <span style="color: #6434A3;">FALSE</span>)
dev.off()
</pre>
</div>


<div id="org58af097" class="figure">
<p><img src="../images/phangorn-demo.png" alt="phangorn-demo.png" width="450px" />
</p>
</div>
</div>
</div>

<div id="outline-container-org66e86f0" class="outline-3">
<h3 id="org66e86f0">Ggtree and friends</h3>
<div class="outline-text-3" id="text-org66e86f0">
<p>
To parse the various tree formats there is the <code>treeio</code> package which has
functions such as <code>read.beast</code> and <code>read.nhx</code>. This will produce a value of the
<code>treedata</code> class which is defined in the <code>tidytree</code> package. The <code>treedata</code> is
an S4 class, so it has <i>slots</i>. So to get the <code>phylo</code> object storing the
topology you would use <code>mytreedata@phylo</code> and to get associated data there is a
<code>data</code> slot.
</p>

<p>
To actually visualise these objects there is the <code>ggtree</code> package which provides
geometries in the style of Ggplot2. Note that to install these packages you will
likely want to use Bioconductor.
</p>

<div class="org-src-container">
<pre class="src src-R">BiocManager::install(<span style="color: #008000;">"ggtree"</span>)
</pre>
</div>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Alexander E. Zarebski</p>
<p class="date">Created: 2024-04-12 Fri 10:29</p>
<p class="validation"><a href="https://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
